---
title: "YXIN_butyrate_feed_PHA"
author: "Pieter van Veelen"
date: "2023-10-25"
output: html_document
---

### Data microbiome profiling data analysis of PHA accumulating biomass 

```{r , eval=TRUE, echo=T, include=T, message=FALSE, warning=FALSE}

# set optional parameters
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE, 
                      message = FALSE)
options(scipen = 999, digits = 3)

```

```{r library loading, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

## load required packages
library(Hmisc)
library(phyloseq)
library(qiime2R)
library(tidyverse)
library(magrittr)
library(devtools)
library(qiime2R)
library(here)
library(breakaway)
library(DivNet)
library(openxlsx)
library(ape)
library(vegan)
library(ggtext)
library(cowplot)
library(RColorBrewer)
library(microbiome)
library(lme4)
library(lmerTest)
library(decontam)
library(ampvis2)
library(glue)
library(lubridate)
library(DECIPHER)
library(ensembleTax)
library(mEQO)
library(ggh4x)

```

```{r project organization, message=F, echo=T, eval=T, warning=F, include=T, cache=T}

# project name
proj = "YXIN_proj1_Q16712"

# create directories
if(!dir.exists("figures")){dir.create("figures")}
if(!dir.exists("output_data")){dir.create("output_data")} 
if(!dir.exists("scripts")){dir.create("scripts")} 
if(!dir.exists("scripts/QIIME2")){dir.create("scripts/QIIME2")} 

### create color vector for later use
library(RColorBrewer)
colorset <- c("darkblue", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "lightskyblue", "darkgreen",  "khaki2", "firebrick", "brown1", "darkorange1", "cyan1", "royalblue4", "darksalmon", "darkblue", "royalblue4", "dodgerblue3", "steelblue1", "lightskyblue", "darkseagreen", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "brown1", "darkorange1", "cyan1", "darkgrey", "darkblue", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "lightskyblue", "darkgreen", "deeppink", "khaki2", "firebrick", "brown1", "darkorange1", "cyan1", "royalblue4", "darksalmon", "darkblue", "royalblue4", "dodgerblue3", "steelblue1", "lightskyblue", "darkseagreen", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "brown1", "darkorange1", "cyan1", "darkgrey")


```


```{r import data, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

physeq = 
qza_to_phyloseq(
  features = "input_data/XYIN_16S_515F_926R_07092023_Q16712_XYIN_table.qza",
  tree = "input_data/XYIN_16S_515F_926R_07092023_Q16712_XYIN_rooted-tree.qza",
  taxonomy = "input_data/XYIN_16S_515F_926R_07092023_Q16712_XYIN_silva-138-99-nb-classifier.qza",
  metadata = "input_data/XYIN_16S_515F_926R_07092023_Q16712_XYIN@metadata_formatted.txt"
  )

# write raw phyloseq object as .rds object
write_rds(physeq, "output_data/physeq.rds")

# check sample metadata and column types
sample_data(physeq)

```

```{bash}

#cp -R /Users/pvee/Wetsus_Projects/BKIS/BKIS_proj1_Q16705/data_analysis/scripts ./scripts

```

```{r clean taxonomy, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

source("scripts/tax_clean_updated.R")

physeq_clean = tax_clean(physeq)

write_rds(physeq_clean, "output_data/physeq_clean_tax.rds")

```

#### Relative proportions of bacterial and non-bacterial sequences
```{r view non-bacterial sequences, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

psdata = physeq_clean
# clean non-informative and non-bacterial ASVs

# quick view on proportion non-bacterial sequences
psdata_nonBac =
psdata %>% 
  tax_glom(., "Phylum") %>% 
  transform_sample_counts(., function(x) x/sum(x)) %>% 
  psmelt %>% 
  as_tibble() %>% 
  select(Sample, Abundance, Kingdom, Phylum) %>% 
  filter(Kingdom != "Bacteria") %>% 
  group_by(Sample, Phylum) %>% 
  summarise(rel_abund = sum(Abundance)) 

psdata_nonBac %>% 
  ggplot(aes(x=Sample, y = rel_abund, fill = Phylum)) +
  geom_col() +
  scale_fill_manual(name=NULL, 
                    values = colorset) + # with named color vector ordered by median Genus abundance
  scale_y_continuous(expand = c(0,0)) +
  labs(x=NULL, y="Relative Abundance (%)") +
  theme_classic() +
  theme(axis.text.x = element_markdown(angle = 45, hjust = 1),
        legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        legend.position = "bottom")


# quick view on proportion Chloroplast and Mitochondrial sequences
psdata_mito_chlorop = 
psdata %>% 
  tax_glom(., "Family") %>% 
  transform_sample_counts(., function(x) x/sum(x)) %>% 
  psmelt %>% 
  as_tibble() %>% 
  filter(Kingdom == "Bacteria") %>% 
  select(Sample, Abundance, Kingdom, Phylum, Family) %>% 
  filter(grepl("Mitochondria|Chloroplast", Family)) %>% 
  group_by(Sample, Family) %>% 
  summarise(rel_abund = sum(Abundance))

psdata_mito_chlorop %>% 
  ggplot(aes(x=Sample, y = rel_abund, fill = Family)) +
  geom_col() +
  scale_fill_manual(name=NULL, 
                    values = colorset) + # with named color vector ordered by median Genus abundance
  scale_y_continuous(expand = c(0,0)) +
  labs(x=NULL, y="Relative Abundance (%)") +
  theme_classic() +
  theme(axis.text.x = element_markdown(angle = 45, hjust = 1),
        legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        legend.position = "bottom")

```

```{r clean on taxonomy, message=F, echo=F, eval=T, warning=F, include=T, cache=F}

# filter unwanted taxonomic groups out
psdata_tax_filtered =
psdata %>% # 5531
  subset_taxa(.,grepl("Bacteria", Kingdom)) %>% # 21 ASVs lost because of other Kingdom than Bacteria
  subset_taxa(., Kingdom != "Unassigned") %>% # 0 ASVs lost
  subset_taxa(., !is.na(Phylum)) %>% # 0 ASVs lost 
  subset_taxa(., Phylum != "Phylum of Bacteria") %>% # 118 ASVs lost
  subset_taxa(., Family != "Mitochondria") %>% # 65 ASVs lost
  subset_taxa(., Class != "Chloroplast") %>% # 0 ASVs lost
  subset_taxa(., Order != "Chloroplast") # 12 ASVs lost


# remove non-informative taxa
old = ntaxa(psdata)
new = ntaxa(psdata_tax_filtered)

# number of ASVs removed:
old-new # 216 ASVs are removed

# check removed Mitochondrial ASV counts per sample
psdata %>% 
  subset_taxa(., Family == "Mitochondria") %>% 
  psmelt %>% 
  select(OTU, Sample, Family,Abundance) %>% 
  mutate_if(is.character, as.factor) %>% 
  group_by(Sample) %>%
summarise(Abundance=sum(Abundance)) %>% 
arrange(desc(Abundance), .by_group = T)

# save unfiltered psdata
psdata_unfiltered = psdata 
saveRDS(object = psdata_unfiltered, file = "output_data/psdata_unfiltered.rds") # save unfiltered data
psdata = psdata_tax_filtered # continue with filtered data as psdata

```

```{r clean phylogeny, message=F, echo=F, eval=T, warning=T, include=T, cache=F}

### resolve phylogenetic tree ###
physeq = physeq_clean

# evaluate tree topology
is.binary(phy_tree(psdata_tax_filtered)) # if FALSE --> polychotomy present (node has more than 2 tips)
#TRUE

# # if FALSE:
# # resolve polychotomous nodes
phy_tree_resolved <- multi2di(phy_tree(physeq))
is.binary(phy_tree_resolved)
# create new phy_tree
tree2 <- phy_tree_resolved

# subset_taxa(phy_tree_resolved, Kingdom ==  "Bacteria")

# merge new phy_tree object with sample_data and otu_table into new phyloseq object
physeq2 <- merge_phyloseq(otu_table(physeq), sample_data(physeq), tax_table(physeq), tree2)

# now filter only the taxa that are clean_filtered by taxonomic filtering
psdata = subset_taxa(physeq2, taxa_names(physeq2) %in% taxa_names(psdata_tax_filtered))

# save filted phyloseq object with resolved tree
saveRDS(object = psdata, file = "output_data/psdata_filtered_tree-resolved.rds")

```

```{r rarefaction curves, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

# alpha rarefaction curve
source("scripts/ampvis2_internals.r")
source("scripts/amp_rankabundance.r")
source("scripts/amp_rarecurve.r")

# show plot
amp_rarecurve(psdata, color = "exp_id", legend.position = "bottomright", xlim = c(0, 120000))

# save plot
pdf(glue("figures/{proj}_rarefaction_curves.pdf"), useDingbats = F, width = 7, height = 5)
amp_rarecurve(psdata, color = "exp_id", xlim = c(0, 120000), legend.position = "bottomright")
dev.off()


```

```{r view mock and blank, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

# extract DNA extraction blank # 216 ASVs
psdata_blank = 
  prune_taxa(taxa_sums(
    subset_samples(psdata, exp_id  == "blank_nc")) > 0,
    subset_samples(psdata, exp_id  == "blank_nc")
    )

# extract mock community # 287 ASVs
psdata_mock = 
  prune_taxa(taxa_sums(
    subset_samples(psdata, exp_id  == "mock")) > 0,
    subset_samples(psdata, exp_id  == "mock")
    )

saveRDS(object = psdata_mock, file = "~/Wetsus_Projects/PVEE/Data_analyses/mock_communities_wetsus/mock_communities_wetsus/input_data/YXIN_proj1_Q16712_PHA_butyrate_feed_mock.rds")

```

```{r remove_mock_and_nc, echo=F}

# keep samples only # 5246 ASVs
psdata_samples = 
  prune_taxa(taxa_sums(
    subset_samples(psdata, !grepl("mock|blank_nc", exp_id))) >0,
    subset_samples(psdata, !grepl("mock|blank_nc", exp_id))
    )

# recreate psdata with samples only
psdata = psdata_samples

```

```{r filter on abundance, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

# relative abundance data
psdata_rel <- transform_sample_counts(psdata, fun = function(x) x/sum(x)) # psdata_rel taxa
ntaxa(psdata_rel)

# total sequence sum
total_sum = sum(sample_sums(psdata)) # 1859672 reads

# continue downstream analysis with abundance filter that retains ASVs with at least 0.1% of total read abundance -> keeps above 99% abundance when removing rare data
psdata_0.01pct <- prune_taxa(taxa_sums(psdata_rel) > 0.0001, psdata) # 3601 taxa (99.7% abundance)
psdata_0.05pct <- prune_taxa(taxa_sums(psdata_rel) > 0.0005, psdata) # 2107 taxa (98.6% abundance)
psdata_0.1pct <- prune_taxa(taxa_sums(psdata_rel) > 0.001, psdata) # 1607 taxa (97.4% abundance)
psdata_1pct <- prune_taxa(taxa_sums(psdata_rel) > 0.01, psdata) # 492 taxa (84.9% abundance)

sum(sample_sums(psdata_0.01pct))/total_sum*100
sum(sample_sums(psdata_0.05pct))/total_sum*100
sum(sample_sums(psdata_0.1pct))/total_sum*100
sum(sample_sums(psdata_1pct))/total_sum*100

psdata_NoAbundanceFilter = psdata
psdata <- psdata_0.05pct # keeping 98.6% of reads in 2107 ASVs

# difference in sample coverage (1.96)
max(sample_sums(psdata_0.05pct)/min(sample_sums(psdata_0.05pct)))

```

```{r prepare data for picrust2, eval=F}

# input data = psdata_0.05pct
biom_test =
psdata_0.05pct %>% 
  otu_table() %>% 
  data.frame(rownames = 1) %>% 
  as_tibble()

biomformat::write_biom(biom_test, glue("output_data/{proj}_otu_table_psdata_0.05pct.biom"))

# try running PICRUST2 as QIIME2 plugin
# DONE

```

#### Agglomeration of ASVs to Genus level groups for downstream analysis for both low N and high N sub-experiments
```{r agglomerate to genus level}

# psdata Genus 
psdata_Genus = tax_glom(
  transform_sample_counts(
    psdata, 
    function(x) x/sum(x)), "Genus")

# psdata Genus Paper 1: Low N conditions
psdata_Genus_lowN = 
  tax_glom(
    prune_taxa(
      taxa_sums(
        subset_samples(psdata, paper_id == "first"))>0, 
        subset_samples(psdata, paper_id == "first")), "Genus")


prune_taxa(taxa_sums(subset_samples(psdata_samples, paper_id == "first"))>0, 
           subset_samples(psdata_samples, paper_id == "first"))

# psdata Genus Paper 2: High N conditions
psdata_Genus_highN = 
  tax_glom(
    prune_taxa(
      taxa_sums(
        subset_samples(psdata, paper_id == "second"))>0, 
        subset_samples(psdata, paper_id == "second")), "Genus")

# create relative abundance phyloseq objects
psdata_Genus_lowN_rel = transform_sample_counts(psdata_Genus_lowN, function(x) x/sum(x))
psdata_Genus_highN_rel = transform_sample_counts(psdata_Genus_highN, function(x) x/sum(x))


# psmelts Genus
ps_Genus = psmelt(psdata_Genus) # n=426
ps_Genus_lowN = psmelt(psdata_Genus_lowN) # n= 426
ps_Genus_highN = psmelt(psdata_Genus_highN) # n= 357

```

#### Write output file usable as input for phaC microbiome-tailored primer design tool (PDT)
```{r create PDT input Genus, echo=F, eval=T}

# create absolute genus physeq
# psdata Genus 
psdata_Genus_abs = tax_glom(psdata, "Genus")

long = 
psdata_Genus_abs %>% 
  psmelt() %>% 
  as_tibble()

# calculate genus-level taxa_sums across samples and calculate taxa prevalence %
wide_taxsums_prev = long %>% 
  select(Sample, Kingdom:Genus, Abundance) %>% 
  pivot_wider(names_from = "Sample", values_from = "Abundance") %>% 
  rowwise() %>% 
  mutate(taxa_sums = sum(c_across(any_of(starts_with("YXIN")))),
         prevalence = mean(c_across(any_of(starts_with("YXIN"))) != 0, na.rm = TRUE) * 100) %>% 
  select(Kingdom:Genus, prevalence, taxa_sums, everything()) 

# make data long again
long_taxsums_prev = 
wide_taxsums_prev %>%
  pivot_longer(cols = any_of(starts_with("YXIN")), names_to = "Sample", values_to = "Abundance")

# calculate sample_sums for sample set
long_sampsums = 
wide_taxsums_prev %>% 
  pivot_longer(cols = any_of(starts_with("YXIN")), names_to = "Sample", values_to = "Abundance") %>% 
  select(Genus, Sample, Abundance) %>% 
  pivot_wider(names_from = "Genus", values_from = "Abundance") %>% 
  rowwise() %>% 
  mutate(sample_sums = sum(c_across(2:ncol(.)))) %>% 
  select(Sample, sample_sums, everything()) %>% 
  pivot_longer(cols = c(3:ncol(.)), names_to = "Genus", values_to = "Abundance")

## PDT data formatted

# which prevalence and abundance thresholds in (%)
# vectors should be of same lengths corresponding to the combinations to make
prevalences = c(80, 90)
abundances = c(0.5, 0.1)

# loop over the data applying combinations of prevalences
for(prev in prevalences) {
  for(abun in abundances) {
    
inner_join(long_sampsums, long_taxsums_prev, by = c("Sample", "Genus")) %>% 
  select(-Abundance.y) %>% 
  rename(Abundance.x = "Abundance") %>% 
  mutate(perc_abund = Abundance/sample_sums*100) %>% 
  select(Sample, Genus, perc_abund, prevalence, "genus_sums" = taxa_sums, Family) %>% 
  group_by(Family, Genus, prevalence, genus_sums) %>% 
  summarise(mean_perc_abund = mean(perc_abund), max_perc_abund = max(perc_abund), min_perc_abund = min(perc_abund), .groups = "drop") %>% 
  filter(prevalence >= prev, mean_perc_abund >= abun) %>% 
  arrange(desc(max_perc_abund)) %>% 
  mutate(`0_Index` = row_number(),
         `1_Name` = if_else(grepl("Genus of ", Genus), Family, Genus),
         `2_Tax` = if_else(grepl("Genus of ", Genus), "family", "genus"),
         `3_mean_genus_perc` = mean_perc_abund,
         `4_perc_samples` = prevalence) %>% 
  select(`0_Index`, `1_Name`, `2_Tax`, `3_mean_genus_perc`, `4_perc_samples`, Family) %>% 
  arrange(desc(`2_Tax`)) %>% 
  write_csv(glue::glue("output_data/{proj}_PDT_input_Genus_prev{prev}_abund{abun}pct.csv"), col_names = T)
  }
}

cat("Input data for PDT saved as output table")

```


```{r alpha diversity rarefy}

source("scripts/avgrarefy.r")

# create subset frequency matrix
YXIN_matrix_lowN <- as(t(otu_table(psdata_Genus_lowN)), "matrix")
YXIN_matrix_highN <- as(t(otu_table(psdata_Genus_highN)), "matrix")

# determine minimal sampling depth
min_sample_lowN <- min(sample_sums(psdata_Genus_lowN)) # 42061
min_sample_highN <- min(sample_sums(psdata_Genus_highN)) # 50627


# rarefaction taking mean of 100 iterations
set.seed(711)
YXIN_matrix_lowN_rare42061_table = avgrarefy(x=YXIN_matrix_lowN, sample = min_sample_lowN, iterations = 100, seed = 711)
YXIN_matrix_highN_rare50627_table = avgrarefy(x=YXIN_matrix_highN, sample = min_sample_highN, iterations = 100, seed = 711)


# create phyloseq object with rarefied data
# low N
psdata_YXIN_lowN_rare <- psdata_Genus_lowN
otu_rare_lowN = otu_table(data.frame(t(YXIN_matrix_lowN_rare42061_table)), taxa_are_rows = TRUE)
otu_table(psdata_YXIN_lowN_rare) <- otu_rare_lowN

# high N
psdata_YXIN_highN_rare <- psdata_Genus_highN
otu_rare_highN = otu_table(data.frame(t(YXIN_matrix_highN_rare50627_table)), taxa_are_rows = TRUE)
otu_table(psdata_YXIN_highN_rare) <- otu_rare_highN

```

### All data are ready for analysis of both papers
Below for now only microbial diversity analysis of Paper 1: low N samples

#### Alpha Diversity
Analysis of genus level alpha diversity based on ASV-level rarefied data
```{r plot alpha diversity rare, message=F, echo=F, eval=T, warning=T, include=T, cache=F}

# calculate alpha diversity: richness & Shannon on Genus level
# input = 
psdata_Genus_lowN

# rarefied
alpha <- estimate_richness(psdata_Genus_lowN, measures = c("Observed", "Chao1", "Shannon", "Simpson"))
alpha$Sample <- row.names(alpha)

metadata = 
  sample_data(psdata_Genus_lowN) %>% 
  data.frame() %>% 
  as_tibble() %>% 
  mutate(Sample = sample_names(psdata_Genus_lowN)) %>% 
  select(Sample, everything()) 

# join alpha diversity with metadata
alpha2 <- inner_join(metadata, alpha, by = "Sample") #%>% 
  #mutate(biofilm = factor(biofilm, levels = c("Broth", "Biofilm")))

# plot alpha diversity
Chao1 <- 
alpha2 %>% 
  mutate(sludge_source = factor(sludge_source, levels = c("Bath", "Almere"))) %>% 
ggplot(aes(x=time, y=Chao1, color = substrate, fill = substrate)) +
  #stat_summary(fun = median, geom = "bar", alpha = 0.3) +
  geom_point(aes(#shape = sampling_day, 
                 fill = substrate,
                 color = substrate,
                 shape = factor(replicate)),
                 #position = position_jitter(width = 0.3)
                 size = 2, show.legend = F) + 
    geom_line(aes(group = interaction(replicate, substrate), 
                 color = substrate),
                 linewidth = 1,alpha=0.5, show.legend = F) + 
  scale_color_manual(values = c("red", "blue")) +
  scale_fill_manual(values = c("red", "blue")) +
  scale_y_continuous(limits = c(150,NA)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(#legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        #axis.ticks.x = element_blank(), 
        strip.background = element_rect(fill="white", colour = "white"),
        axis.title.y = element_text(hjust = 0.5),
        strip.placement = "outside",
        legend.position = "bottom",
        panel.border = element_rect(colour = "black", fill = NA)) +
  labs(y = expression("Genus Chao1"),
       x = NULL) +
  theme(axis.text.x = element_text(hjust = 0.5)) +
  ggrepel::geom_text_repel(aes(label = ifelse(replicate == 1 & time == "48h", substrate, "")), 
                           size = 3, force_pull = 0, nudge_y = -10, nudge_x = 0.1, min.segment.length = 10, show.legend = F) +
  theme(strip.text = element_text(face = "bold", size = 12, hjust = 0)) +
  facet_nested(~ sludge_source, scales = "free", space = "free_x")


# print Chao1 Genus
print(Chao1)



# plot alpha diversity
Shannon <- 
alpha2 %>% 
  mutate(sludge_source = factor(sludge_source, levels = c("Bath", "Almere"))) %>% 
ggplot(aes(x=time, y=Shannon, color = substrate, fill = substrate)) +
  #stat_summary(fun = median, geom = "bar", alpha = 0.3) +
  geom_point(aes(shape = factor(replicate), 
                 fill = substrate,
                 color = substrate)
                 #position = position_jitter(width = 0.3)
                 , size = 2, show.legend = T) + 
    geom_line(aes(group = interaction(replicate, substrate), 
                 color = substrate)
                 , linewidth = 1,alpha=0.5, show.legend = F) + 
  scale_color_manual(values = c("red", "blue")) +
  scale_fill_manual(values = c("red", "blue")) +
  scale_shape_discrete(name = "replicate",) +
  #scale_y_continuous(limits = c(0,NA)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(#legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        #axis.ticks.x = element_blank(), 
        strip.background = element_rect(fill="white", colour = "white"),
        axis.title.y = element_text(hjust = 0.5),
        strip.placement = "outside",
        legend.position = "bottom",
        panel.border = element_rect(colour = "black", fill = NA)) +
  labs(y = expression("Shannon diversity H' (Genus level)"),
       x = NULL) +
  theme(axis.text.x = element_text(hjust = 0.5)) +
  ggrepel::geom_text_repel(aes(label = ifelse(replicate == 1 & time == "48h", substrate, "")), 
                           size = 3, force_pull = 0, nudge_y = 0.3, nudge_x = 0.1, min.segment.length = 10,
                           show.legend = F) +
  theme(strip.text = element_text(face = "bold", size = 12, hjust = 0)) +
  facet_nested(~ sludge_source, scales = "free", space = "free_x")

print(Shannon)

# create plot
prow = 
  plot_grid(
  plot_grid(Chao1 + theme(legend.position = "none"), 
            Shannon + theme(legend.position = "none"),
                   align = "hv",
                   labels = c("A", "B"),
                   hjust = -1, 
                   nrow = 1),
  get_legend(Chao1), 
  nrow = 2 , 
  rel_heights = c(5, 1))


# show plot
prow

# save plot
ggsave(prow, filename = glue("figures/{proj}_plot_alpha_div.pdf"), width = 7, height = 5)
 
# save Shannon plot
ggsave(Shannon, filename = glue("figures/{proj}_plot_alpha_div.pdf"), width = 7, height = 4)

```

```{r alpha statistics}

model_alpha = 
alpha2 %>% 
  mutate(sludge_source = factor(sludge_source, levels = c("Bath", "Almere"))) %>% 
  with(lm(Shannon ~ sludge_source + substrate * time)) 

# anova result
anova(model_alpha)

# test differential effects of time on Shannon between substrates and sludge
emmeans_shannon = emmeans::emmeans(model_alpha, pairwise ~ time | substrate + sludge_source, p.adjust="Holm")

# save anova contrasts
emmeans_shannon$contrasts %>% 
  as_tibble() %>% 
  write_csv(glue::glue("output_data/{proj}_anova_result_Shannon_Tukey_pvals.csv"))

```

```{r write genus tables, echo=F}

# paper 1 only
### write rel abund data plus other tax info
ps_Genus_lowN %>% 
  # filter(!grepl("mock", combined), # remove mock communities
  #        Sample_mock != "blank") %>% # remove NC blanks
  as_tibble() %>% 
  mutate(across(everything(),  ~str_replace_all(., ",", "."))) %>% 
  mutate(across(where(~ all(grepl("^-?\\d*\\.?\\d*$", .))), as.numeric)) %>% 
  select(Sample, OTU, Abundance, Kingdom:Genus, everything()) %>% 
  group_by_at(vars(Sample, Kingdom:Genus)) %>% 
  summarise(rel_abund = sum(Abundance), .groups = "drop") %>% 
  group_by_at(vars(-rel_abund)) %>% 
  summarise(mean_rel_abund = 100* mean(rel_abund), .groups = "drop") %>% 
  select(Kingdom:Genus, Sample, mean_rel_abund) %>% 
  pivot_wider(names_from = Sample, values_from = mean_rel_abund, values_fill = 0) %>% 
  group_by_at(vars(Kingdom:Genus)) %>% 
  rowwise() %>% 
  mutate(mean = mean(c_across(where(is.numeric)))) %>% 
  arrange(desc(mean)) %>% 
  select(-mean) %>% 
  ungroup() %>% 
  mutate(across(where(is.numeric), ~round(.x, digits = 2))) %>% 
  select(Kingdom:Genus, everything()) %>% view
  write_csv(glue::glue("output_data/{proj}_relative_Genus_abundances_sorted_by_overall_mean.csv", col_names = T, append = F))

```

#### Barplots of genus level taxonomy
```{r plot_bar all genera, echo=FALSE, fig.width=8, fig.height=10}

# paper 1 only
#ps_Genus_lowN

# count n genera n = 426
ps_Genus_lowN %>% 
  as_tibble() %>% 
  mutate(Genus = factor(Genus)) %>% 
  pull(Genus) %>% 
  levels() %>% 
  length()

# Define a function to find the first non-matching column

# Use rowwise and mutate to apply the function
ps_Genus_taxlabel = ps_Genus_lowN %>% 
  as_tibble() %>% 
  mutate(across(everything(),  ~str_replace_all(., ",", "."))) %>% 
  mutate(across(where(~ all(grepl("^-?\\d*\\.?\\d*$", .))), as.numeric)) %>% 
  select(Sample, Abundance, paper_id:Genus) %>%
  rowwise() %>%
  mutate(tax_label = case_when(
    grepl("\\d", Genus) ~ {
      first_non_numeric <- case_when(
        !grepl("\\d", Family) ~ paste(Family, Genus, sep = " "),
        !grepl("\\d", Order) ~ paste(Order, Genus, sep = " "),
        !grepl("\\d", Class) ~ paste(Class, Genus, sep = " "),
        !grepl("\\d", Phylum) ~ paste(Phylum, Genus, sep = " "),
        !grepl("\\d", Kingdom) ~ paste(Kingdom, Genus, sep = " "),
        TRUE ~ NA_character_
      )
      first_non_numeric
    },
    TRUE ~ Genus
  ))

Genus_abundances <- ps_Genus_taxlabel %>% 
  mutate(Sample_nr = str_sub(Sample, start = 1, end = 2)) %>% 
  group_by_at(vars(Sample, Sample_nr, sludge_source, substrate, replicate,exp_id, time, tax_label)) %>% 
  summarise(rel_abund = sum(Abundance), .groups = "drop") %>%
  group_by(Sample) %>% 
  mutate(mean_rel_abund = rel_abund/sum(rel_abund)*100) %>% 
  ungroup() %>% 
  group_by(Sample, tax_label) %>% 
  mutate(tax_label = str_replace(tax_label, "(.*)_unclassified", "Unclassified *\\1*"),
         tax_label = str_replace(tax_label, "^(\\S*)$", "*\\1*"))

# Reduce genera to lower number (at least 3%, or else called other)
Genus_pool <- Genus_abundances %>% 
  group_by(tax_label) %>% 
  summarise(pool = max(mean_rel_abund) < 3, 
            mean = mean(mean_rel_abund), 
            .groups = "drop")

# define a color vector using brewer.pal and colorset
colvec = c(brewer.pal(8,"Dark2"),brewer.pal(8,"Set2"),brewer.pal(8,"Accent"),brewer.pal(8,"Paired"),brewer.pal(12,"Set3"),brewer.pal(9,"Set1"), brewer.pal(9,"YlGnBu"), colorset)

# name the taxon colors
inner_join(Genus_abundances, Genus_pool, by="tax_label") %>% 
  mutate(tax_label = if_else(pool, "Other max.<3%", tax_label)) %>% 
  group_by(tax_label) %>%
  summarise(mean_rel_abund = sum(mean_rel_abund), 
            median = median(mean),
            .groups = "drop") %>% 
  mutate(tax_label = factor(tax_label), 
         tax_label = fct_reorder(tax_label, median, .desc = T)) %>% 
  arrange(desc(median)) %>% pull(tax_label) %>% unique() -> names(colvec)

colvec["Other max.<3%"] <- "#D3D3D3"

# reshape plot data
plot_data <-
inner_join(Genus_abundances, Genus_pool, by="tax_label") %>% 
  mutate(tax_label = if_else(pool, "Other max.<3%", tax_label)) %>% 
  # mutate(Sample = if_else(grepl(".*[^0-9][0-9]{2}$", Sample),
  #                          Sample,
  #                          str_c(str_sub(Sample, end = -2),"0", str_sub(Sample, -1)))) %>% 
  group_by(Sample, Sample_nr, sludge_source, replicate, substrate, time, tax_label) %>% 
  summarise(mean_rel_abund = sum(mean_rel_abund), 
            mean = min(mean),
            .groups = "drop") %>% 
  mutate(tax_label = factor(tax_label), 
         tax_label = fct_reorder(tax_label, mean, .desc = T)) 



# plot the data per sample
plot_Genus = 
  plot_data %>% 
ggplot(aes(x=time, y = mean_rel_abund, fill = tax_label)) +
  geom_col() +
  scale_fill_manual(name=NULL, 
                    values = colvec) + # with named color vector ordered by median Genus abundance
  scale_y_continuous(expand = c(0,0)) +
  labs(x=NULL, y="Relative Abundance (%)") +
  theme_classic() +
  theme(axis.text.x = element_markdown(angle = 0, hjust = 0.5),
        legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        legend.position = "bottom",
        strip.background = element_rect(colour = "white"),
        strip.text = element_text(face = "bold")) +
  guides(fill = guide_legend(nrow=12))
  

# print plot  
gg_Genus = plot_Genus + facet_nested_wrap(~sludge_source + replicate + substrate, ncol=4,  scales = "free")
gg_Genus

# save plots
ggsave(plot = gg_Genus, filename = glue("figures/{proj}_barplot_all_Genus_reactorsOnly.pdf"), width = 8, height = 6)

```

```{r plot_bars_example_for_paper}

# plot data per timepoint, average over replicates
average_data = 
  plot_data %>% 
  group_by(sludge_source, substrate, time, tax_label) %>% 
  summarise(mean_rel_abund = mean(mean_rel_abund), sd = sd(mean_rel_abund)) 

# plot data for rep1 only
rep1_data = 
  plot_data %>% 
  filter(replicate == "1")


plot_bars = function(data)  {
  
plot =   
ggplot(data = data, aes(x=time, y = mean_rel_abund, fill = tax_label)) +
  geom_col() +
  scale_fill_manual(name=NULL, 
                    values = colvec) + # with named color vector ordered by median Genus abundance
  scale_y_continuous(expand = c(0,0)) +
  labs(x=NULL, y="Relative Abundance (%)") +
  theme_classic() +
  theme(axis.text.x = element_markdown(angle = 0, hjust = 0.5),
        legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        legend.position = "bottom",
        strip.background = element_rect(colour = "white"),
        strip.text = element_text(face = "bold")) +
  guides(fill = guide_legend(nrow=12)) + 
  facet_nested_wrap(~sludge_source + substrate, 
                    ncol=4,  scales = "free")

return(plot)
}  

average_plot = plot_bars(data = average_data)
rep1_plot = plot_bars(data = rep1_data)

# save plots
ggsave(plot = average_plot, 
       filename = glue("figures/{proj}_barplot_replicate_average_Genus.pdf"), width = 8, height = 4)
ggsave(plot = rep1_plot,
       filename = glue("figures/{proj}_barplot_replicate_replicate1_Genus.pdf"), width = 8, height = 4)

```


```{r geom_tile, echo = F, fig.width=10, fig.height=10}

# remarks order for x axis
#order = c(1,19,16,6,5,9,17,7,8,4,3,18,2,10,11,12,13,14,15)

taxon_pool <- Genus_abundances %>% 
  group_by(tax_label) %>% 
  summarise(pool = max(mean_rel_abund) < 3, 
            mean = mean(mean_rel_abund), 
            .groups = "drop")

heatmap_data_long = 
Genus_abundances %>% 
  group_by(Sample,sludge_source, replicate, substrate,time, tax_label) %>% 
  summarise(mean_rel_abund = mean(mean_rel_abund), .groups = "drop") %>% 
inner_join(., Genus_pool, by="tax_label") %>% 
  group_by(Sample,sludge_source, replicate, substrate,time, tax_label) %>% 
  mutate(tax_label = if_else(pool, "Other max.<3%", tax_label)) %>% 
  select(Sample, sludge_source, replicate, substrate, time,tax_label, mean_rel_abund, mean, pool) %>% 
  group_by(Sample, sludge_source, replicate, substrate, time,tax_label) %>% 
  summarise(sum_rel_abund = sum(mean_rel_abund)) %>% 
  mutate(tax_label = factor(tax_label), 
         tax_label = fct_reorder(tax_label, sum_rel_abund, .desc = F)) %>%
  ungroup()

# determine order of taxa on y axis from mean high to low
tax_order = 
heatmap_data_long %>% 
  mutate(tax_label = as.character(tax_label)) %>% 
  group_by(tax_label) %>% 
  summarise(mean = mean(sum_rel_abund), .groups = "drop") %>% 
  arrange(desc(mean)) %>% pull(tax_label)

# create wide format data
heatmap_data = 
heatmap_data_long %>% 
select(Sample, sludge_source, replicate, substrate,time, tax_label,time, sum_rel_abund) %>% 
  pivot_wider(names_from = tax_label, 
              values_from = sum_rel_abund) 

# make heatmap
m <- as.matrix(heatmap_data[, 6:ncol(heatmap_data)])
clust <- hclust(dist(t(m)))


# direction of taxon changes over time
taxon_changes = 
heatmap_data_long %>% 
  mutate(tax_label = fct_relevel(tax_label, rev(tax_order))) %>%
  select(-Sample) %>% 
  pivot_wider(names_from = "time", values_from = "sum_rel_abund", values_fill = 0) %>% 
  mutate(time_diff = 
           case_when(`24h`-`0h` > 0 & `48h`-`0h` > 0  ~ "increasing", 
                     `24h`-`0h` < 0 & `48h`-`0h` < 0  ~ "decreasing",
                     TRUE ~ "stable")) %>% 
  select(sludge_source, substrate, replicate, tax_label, time_diff) %>% 
  pivot_wider(names_from = c("replicate"), values_from = "time_diff") %>% 
  mutate(direction = case_when(`1` == "increasing" & `2` == "increasing" ~ "increasing",
                               `1` == "decreasing" & `2` == "decreasing" ~ "decreasing",
                               `1` == "stable" & `2` == "stable" ~ "stable",
                               `1` != `2` ~ "inconsistent",
                               TRUE ~ "other")) %>% 
  select(sludge_source, substrate, tax_label, direction) %>% 
  pivot_wider(names_from = substrate, values_from = direction, values_fill = NA_character_) %>% 
  mutate(change = case_when(butyrate == "increasing" & acetate == "increasing" ~ "general_increase",
                                      butyrate == "increasing" & acetate != "increasing" ~ "butyrate_increase",
                                      butyrate != "increasing" & acetate == "increasing" ~ "acetate_increase",
                                      butyrate == "decreasing" & acetate == "decreasing" ~ "general_decrease",
                                      butyrate == "decreasing" & acetate != "decreasing" ~ "butyrate_decrease",
                                      butyrate != "decreasing" & acetate == "decreasing" ~ "acetate_decrease",
                                      butyrate == "inconsistent" & acetate == "inconsistent" ~ "general_inconsistent",
                                      butyrate == "inconsistent" & acetate != "inconsistent" ~ "butyrate_inconsistent",
                                      butyrate != "inconsistent" & acetate == "inconsistent" ~ "acetate_inconsistent",
                                      TRUE ~ "other")) %>% 
  select(sludge_source, tax_label, change)
                               
                            
### TODO WITH THE INCREASE/DECREASE/STABLE CATEGORIZATION OF THE HEATMAP
# plot heatmap
heatmap_genus_3pct = 
heatmap_data_long %>% 
  mutate(tax_label = fct_reorder(tax_label, sum_rel_abund)) %>% 
  left_join(., taxon_changes, by = c("sludge_source","tax_label")) %>% 
  filter(tax_label != "Other max.<3%",
         #sludge_source == "Almere"
         ) %>% 
  mutate(change = factor(change, levels = c("butyrate_increase", "butyrate_decrease", "butyrate_inconsistent",
                                            "acetate_increase", "acetate_decrease", "acetate_inconsistent",
                                            "general_increase", "general_decrease", "general_inconsistent")),
         dir = str_remove(change, "^(.*)_"),
         dir = factor(dir, levels = c("increase", "decrease", "inconsistent")),
         which = str_remove(change, "_(.*)"),
         which = factor(which, levels = c("butyrate", "acetate", "general"))) %>% 
ggplot(aes(time, tax_label)) +
  geom_tile(aes(fill = sum_rel_abund), color = NA) +
  labs(x = NULL, y=NULL) +
  #scale_y_discrete(limits = colnames(m)[rev(clust$order)]) +
  scale_fill_gradient(low = "white", high = "darkred",
                      name = "Relative\nAbundance (%)\n(\u00B7 0%)") +
  scale_x_discrete(position = "top") +
  theme(axis.text.x = element_text(size = 8, angle = 90, hjust = 1, vjust = -1),
        strip.text.x = element_text(angle = 0),
        strip.text.y = element_text(angle = 0),
        axis.ticks.y = element_blank(),
        axis.ticks.x = element_line(colour = "grey70"),
        axis.text.y = element_markdown(size = 8),
        plot.background = element_blank(),
        strip.placement = "outside",
        panel.border = element_rect(colour = "grey90", fill = NA),
        panel.grid = element_blank(),
        panel.background = element_rect(fill = "grey90")) +
  geom_text(
  aes(label = ifelse(sum_rel_abund >3 & sum_rel_abund < 17, round(sum_rel_abund, digits = 0), "")),
  color = "grey50", size =2.5) +
  geom_text(
  aes(label = ifelse(sum_rel_abund == 0, ".", "")),
  color = "grey80", size = 5, position = position_nudge(y=0.5)) +
  geom_text(
  aes(label = ifelse(sum_rel_abund > 17, round(sum_rel_abund, digits = 0), "")),
  color = "grey90", size = 2.5) +
  facet_nested(rows = vars(which, dir), 
               cols = vars(sludge_source, substrate,replicate), space = "free", scales = "free")

 print(heatmap_genus_3pct)

ggsave2(plot = heatmap_genus_3pct + 
                ggtitle(label = "Genus heatmap -\nYXIN: PHA - acetate vs. butyrate under low N", 
                        subtitle = "Genera with abundance >3% in at least one sample\nPercent abundance shown in cells when >3%"), 
        filename = glue("figures/{proj}_heatmap_genus_3pct.pdf"), width = 10, height = 8)

# save plot data with max.<3% others
heatmap_data_long %>% 
  mutate(tax_label = fct_relevel(tax_label, rev(tax_order))) %>% 
  select(Sample, tax_label, sum_rel_abund) %>%
  pivot_wider(names_from = "Sample", values_from = "sum_rel_abund", values_fill = 0) %>%
write.xlsx(., file = glue("output_data/{proj}_heatmap_data_genus_table_3pct.xlsx"), col_names = T)

# save plot data without other category
ps_Genus %>% 
  as_tibble() %>% 
  select(sludge_source, replicate, substrate, Sample,time, Abundance, Kingdom:Genus) %>% 
  group_by_at(vars(sludge_source, replicate, substrate, Sample,time, Kingdom:Genus)) %>% 
  summarize(rel_abund = sum(Abundance), .groups = "drop") %>% 
  group_by_at(vars(Sample, Kingdom:Genus)) %>% 
  summarize(mean_rel_abund = 100* mean(rel_abund), .groups = "drop") %>% 
  mutate(Genus = str_replace(Genus, "(.*)_unclassified", "Unclassified *\\1*"),
         Genus = str_replace(Genus, "^(\\S*)$", "*\\1*")) %>% 
  pivot_wider(names_from = "Sample", values_from = "mean_rel_abund", values_fill = 0) %>%
  write.xlsx(., file = glue("output_data/{proj}_heatmap_data_genus_table_all.xlsx"), col_names = T) 
  
```

```{r split_heatmap}

# create separate heatmaps of only increasing and decreasing taxa
# create separate figures for acetate and butyrate feeds, and separated for Almere and Bath as well
# exclude the inconsistent/stable group ot taxa

heatmap_data_long_for_split =
  heatmap_data_long %>% 
  mutate(tax_label = fct_reorder(tax_label, sum_rel_abund)) %>% 
  left_join(., taxon_changes, by = c("sludge_source","tax_label")) %>% 
  filter(tax_label != "Other max.<3%") %>% 
  mutate(change = factor(change, levels = c("butyrate_increase", "butyrate_decrease", "butyrate_inconsistent",
                                            "acetate_increase", "acetate_decrease", "acetate_inconsistent",
                                            "general_increase", "general_decrease", "general_inconsistent")),
         dir = str_remove(change, "^(.*)_"),
         dir = factor(dir, levels = c("increase", "decrease", "inconsistent")),
         which = str_remove(change, "_(.*)"),
         which = factor(which, levels = c("butyrate", "acetate", "general"))) %>% 
  filter(dir != "inconsistent")



# Create and save individual plots

# Initialize an empty list to store ggplot objects
plot_list <- list()
 
for (level1 in unique(heatmap_data_long_for_split$sludge_source)) {
  for (level2 in unique(heatmap_data_long_for_split$substrate)) {
    subset_data <- 
      heatmap_data_long_for_split %>% 
      filter(grepl(level1, sludge_source) & grepl(level2, substrate)) %>% 
      filter(which %in% c("general", level2))
    
# Create ggplot for each subset    
p <- 
  subset_data %>% 
  mutate(tax_label = fct_reorder(tax_label, sum_rel_abund)) %>% 
ggplot(aes(time, tax_label)) +
  geom_tile(aes(fill = sum_rel_abund), color = NA) +
  labs(x = NULL, y=NULL) +
  #scale_y_discrete(limits = colnames(m)[rev(clust$order)]) +
  scale_fill_gradient(low = "white", high = "darkred",
                      limits = c(0, 35),
                      name = "Relative\nAbundance (%)\n(\u00B7 0%)") +
  scale_x_discrete(position = "top") +
  theme(axis.text.x = element_text(size = 8, angle = 90, hjust = 1, vjust = -1),
        strip.text.x = element_text(angle = 0),
        strip.text.y = element_text(angle = 0),
        axis.ticks.y = element_blank(),
        axis.ticks.x = element_line(colour = "grey70"),
        axis.text.y = element_markdown(size = 8),
        plot.background = element_blank(),
        strip.placement = "outside",
        panel.border = element_rect(colour = "grey90", fill = NA),
        panel.grid = element_blank(),
        panel.background = element_rect(fill = "grey90")) +
  geom_text(
  aes(label = ifelse(sum_rel_abund >3 & sum_rel_abund < 17, round(sum_rel_abund, digits = 0), "")),
  color = "grey50", size =2.5) +
  geom_text(
  aes(label = ifelse(sum_rel_abund == 0, ".", "")),
  color = "grey80", size = 5, position = position_nudge(y=0.5)) +
  geom_text(
  aes(label = ifelse(sum_rel_abund > 17, round(sum_rel_abund, digits = 0), "")),
  color = "grey90", size = 2.5) +
  facet_nested(rows = vars(which, dir), 
               cols = vars(sludge_source, 
                           substrate,
                           factor(replicate, labels =c("replicate 1", "replicate2"))), 
               space = "free", scales = "free") +
  #ggtitle(paste0("Sludge origin ",level1," - ", level2, " substrate")) +
theme(
  plot.title = element_text(hjust = 0.1, vjust = 0),
  panel.background = element_rect(fill='transparent'),
  panel.border = element_rect(colour = "black", fill=NA, linewidth = 0.3),
  axis.line.x = element_line(color="black", size = 0.3),
  axis.line.y = element_line(color="black", size = 0.3),
  strip.background = element_rect(fill = "transparent", color = "black", linewidth = 0.3),
  strip.text = element_text(size = 8)
  )

  # # Save each plot as a separate file
     ggsave(plot = p, filename = glue::glue("figures/{proj}_plot_",level1,"_",level2,".pdf"))
  
  # Remove legend from all plots except the first one
    if (length(plot_list) == 0) {
      p <- p + theme(legend.position = "bottom")
    } else {
      p <- p + theme(legend.position = "none")
    }
     
    # Save each ggplot object to the list
    plot_list[[length(plot_list) + 1]] <- p
  }
}

# Combine all ggplot objects using patchwork
combined_plot <- patchwork::wrap_plots(plotlist = plot_list, ncol = 4)
combined_plot

# Save or display the combined plot
ggsave(plot = combined_plot, filename = glue::glue("figures/{proj}_comined_heatmap_plot.pdf"), width = 20, height = 7)
                                       

```

### Beta diversity

```{r total beta diversity rel_abund, message=F, fig.width=15, fig.height=6, echo=F, eval=T, warning=F, include=T, cache=F}

# recode days as categorical
colnames(sample_data(psdata_Genus_lowN))

### Beta diversity analysis first
# input data  = psdata
psdata_Genus_lowN_rel = transform_sample_counts(psdata, function(x) x/sum(x))
psdata_Genus_lowN_clr = microbiome::transform(psdata_Genus_lowN, transform = "clr")

sample_data(psdata_Genus_lowN_rel_v2)$replicate = factor(sample_data(psdata_Genus_lowN_rel_v2)$replicate)
sample_data(psdata_Genus_lowN_rel_v2)$sludge_source = factor(sample_data(psdata_Genus_lowN_rel_v2)$sludge_source, levels = c("Bath", "Almere"))


# ordination
PCoA_BC <- ordinate(psdata_Genus_lowN_rel_v2, method = "PCoA", distance = "bray")
PCoA_Jac <- ordinate(psdata_Genus_lowN_rel_v2, method = "PCoA", distance = "jaccard")
PCoA_uu <- ordinate(psdata_Genus_lowN_rel_v2, method = "PCoA", distance = "uunifrac")
PCoA_wu <- ordinate(psdata_Genus_lowN_rel_v2, method = "PCoA", distance = "wunifrac")

# nmds_BC <- ordinate(psdata_Genus_lowN_rel_v2, method = "NMDS", distance = "bray")
# nmds_Jac <- ordinate(psdata_Genus_lowN_rel_v2, method = "NMDS", distance = "jaccard")
# nmds_uu <- ordinate(psdata_Genus_lowN_rel_v2, method = "NMDS", distance = "uunifrac")
# nmds_wu <- ordinate(psdata_Genus_lowN_rel_v2, method = "NMDS", distance = "wunifrac")

plot_PCoA_BC <- 
  plot_ordination(physeq = psdata_Genus_lowN_rel_v2, 
                  ordination = PCoA_BC, 
                  type = "samples", 
                  axes = c(1,2), 
                  color = "time", shape = "replicate") + 
  ggrepel::geom_text_repel(aes(label = time), size = 2) +
  scale_color_manual(values = c(brewer.pal(6, "Dark2"))) + 
  geom_point(size=2) +
  #ggtitle("Bray Curtis\n(presence and abundance)") + 
  facet_nested(~sludge_source+substrate)
  

plot_PCoA_Jac <- 
  plot_ordination(physeq = psdata_Genus_lowN_rel_v2, 
                  ordination = PCoA_Jac, 
                   type = "samples", 
                  axes = c(1,2), 
                  color = "time", shape = "replicate") + 
  ggrepel::geom_text_repel(aes(label = time), size = 2) +
  scale_color_manual(values = c(brewer.pal(6, "Dark2"))) + 
  geom_point(size=2) +
  ggtitle("Jaccard\n(binary presence only)") +
  facet_nested(~sludge_source + substrate)

plot_PCoA_uu <- 
  plot_ordination(physeq = psdata_Genus_lowN_rel_v2, 
                  ordination = PCoA_uu, 
                   type = "samples", 
                  axes = c(1,2), 
                  color = "time", shape = "replicate") + 
  ggrepel::geom_text_repel(aes(label = time), size = 2) +
  scale_color_manual(values = c(brewer.pal(6, "Dark2"))) + 
  geom_point(size=2) +
  ggtitle("unweighted UniFrac \n(binary lineage presence only)") +
  facet_nested(~sludge_source + substrate)


# Combine the plots with shared legend
combined_plots <- patchwork::wrap_plots(#plot_PCoA_Jac, 
                                        #plot_PCoA_uu
                                        plot_PCoA_BC +
theme(
  plot.title = element_text(hjust = 0.1, vjust = -10),
  panel.background = element_rect(fill='transparent'),
  panel.grid.major = element_line(colour = "grey90"),
  strip.text = element_text(face = "bold"),
  panel.border = element_rect(colour = "black", fill=NA, linewidth = 0.3),
  axis.line.x = element_line(color="black", size = 0.3),
  axis.line.y = element_line(color="black", size = 0.3))
) +
  patchwork::plot_layout(guides = "collect", ncol = 1) 

# Display the combined plot
combined_plots

ggsave(plot = combined_plots, filename = glue("figures/{proj}_plot_beta_diversity_all_rel.pdf"), width = 9, height =3)


```

```{r PERMANOVA statistics}

psdata_Genus_lowN_rel = tax_glom(psdata_Genus_lowN_rel_v2, "Genus")
sample_data(psdata_Genus_lowN_rel)$substrate = factor(sample_data(psdata_Genus_lowN_rel)$substrate)

# sample data
sam_data <- data.frame(sample_data(psdata_Genus_lowN_rel))
asv_data <- vegdist(t(otu_table(psdata_Genus_lowN_rel)), method = "bray")
# calculate adonis permanova on granules only on ASV level Bray-Curtis dissimilarity (tested within sludge source)
adonis_bc = adonis2(asv_data ~  sludge_source * substrate * time, 
                    permutations = 999, 
                    strata = sam_data$sludge_source, 
                    data = sam_data)

# view PERMANOVA output
adonis_bc

# test beta dispersion assumption (no significant deviations)
betadisper_bc <- betadisper(asv_data, with(sam_data, interaction(sludge_source, substrate)))
permutest(betadisper_bc,  pairwise=T)    


```



### Beta diversity per replicate run
```{r total beta diversity rel_abund per replicate, fig.width=20, fig.height=6, message=F, echo=F, eval=T, warning=F, include=F, cache=F}

# input data
psdata_Genus_lowN_rel_v2 = subset_samples(psdata_Genus_lowN_rel, condition == "Low_N")

# split by replicate
# list objects
physeq_list_reps = c(
Almere_1_ac =   
    subset_samples(psdata_Genus_lowN_rel_v2, 
                   sludge_source == "Almere" & replicate == 1 & substrate == "acetate"),
Almere_2_ac =   
    subset_samples(psdata_Genus_lowN_rel_v2, 
                   sludge_source == "Almere" & replicate == 2 & substrate == "acetate"),
Bath_1_ac = 
    subset_samples(psdata_Genus_lowN_rel_v2, 
                   sludge_source == "Bath" & replicate == 1 & substrate == "acetate"),
Bath_2_ac = 
    subset_samples(psdata_Genus_lowN_rel_v2, 
                   sludge_source == "Bath" & replicate == 2 & substrate == "acetate"),
Almere_1_but = 
    subset_samples(psdata_Genus_lowN_rel_v2, 
                   sludge_source == "Almere" & replicate == 1 & substrate == "butyrate"),
Almere_2_but = 
    subset_samples(psdata_Genus_lowN_rel_v2, 
                   sludge_source == "Almere" & replicate == 2 & substrate == "butyrate"),
Bath_1_but = 
    subset_samples(psdata_Genus_lowN_rel_v2, 
                   sludge_source == "Bath" & replicate == 1 & substrate == "butyrate"),
Bath_2_but = 
    subset_samples(psdata_Genus_lowN_rel_v2, 
                   sludge_source == "Bath" & replicate == 2 & substrate == "butyrate")
)


# Run ordination on each phyloseq object in the list
ordination_list <- lapply(physeq_list_reps, function(physeq) {
  ord_result_BC <- ordinate(physeq, method = "PCoA", distance = "bray")
  ord_result_Jac <- ordinate(physeq, method = "PCoA", distance = "jaccard")
  ord_result_uu <- ordinate(physeq, method = "PCoA", distance = "uunifrac")
  ord_result_wu <- ordinate(physeq, method = "PCoA", distance = "wunifrac")
  return(list(BC = ord_result_BC, Jac = ord_result_Jac, UU = ord_result_uu, WU = ord_result_wu))
})

# plot all combinatorial sets
plot_ordination_from_list <- function(ordination_list, metric) {
  plots <- lapply(names(ordination_list), function(name) {
    ord_result <- ordination_list[[name]][[metric]]
    plot_ordination(ordination = ord_result, 
                    physeq = physeq_list_reps[[name]],
                    color = "time", 
                    shape = "substrate",
                    title = paste(name, metric, sep = " - ")) +
      scale_y_continuous(limits = c(-0.4, 0.4)) +
      scale_x_continuous(limits = c(-0.45, 0.45)) +
      geom_point(size = 3)
  })
  return(plots)
}

# Plot each ordination using the specified metric
metric_to_plot <- "Jac"  # Change this to "BC", "Jac", "UU", or "WU"
plots <- plot_ordination_from_list(ordination_list, metric_to_plot)
gridExtra::grid.arrange(grobs = plots, ncol = 4)

# Display the plots
ggsave(plot = gridExtra::grid.arrange(grobs = plots, ncol = 4), filename =  glue("figures/{proj}_{metric_to_plot}_plots_individual_groups.pdf"), width = 20, height = 7)

# Plot each ordination using the specified metric
metric_to_plot <- "BC"  # Change this to "BC", "Jac", "UU", or "WU"
plots <- plot_ordination_from_list(ordination_list, metric_to_plot)
#gridExtra::grid.arrange(grobs = plots, ncol = 4)

# Display the plots
ggsave(plot = gridExtra::grid.arrange(grobs = plots, ncol = 4), filename =  glue("figures/{proj}_{metric_to_plot}_plots_individual_groups.pdf"), width = 20, height = 7)


```

```{r beta diversity distances from a single ordination}

phylist = list(physeq_all = psdata_Genus_lowN_rel_v2)

# Run ordination on each phyloseq object in the list
ordination_list <- lapply(phylist, function(physeq) {
  ord_result_BC <- ordinate(physeq, method = "PCoA", distance = "bray")
  ord_result_Jac <- ordinate(physeq, method = "PCoA", distance = "jaccard")
  ord_result_uu <- ordinate(physeq, method = "PCoA", distance = "uunifrac")
  ord_result_wu <- ordinate(physeq, method = "PCoA", distance = "wunifrac")
  return(list(BC = ord_result_BC, Jac = ord_result_Jac, UU = ord_result_uu, WU = ord_result_wu))
})

# plot all combinatorial sets
plot_ordination_from_list <- function(ordination_list, metric, fix_axes = F) {
  plots <- lapply(names(ordination_list), function(name) {
    ord_result <- ordination_list[[name]][[metric]]
    if(fix_axes == TRUE){
    plot_ordination(ordination = ord_result, 
                    physeq = phylist[[name]], # ADJUST TO PHYLOSEQ list name
                    color = "time", 
                    shape = "substrate",
                    title = paste(name, metric, sep = " - ")) +
      scale_y_continuous(limits = c(-0.3, 0.3)) +
      scale_x_continuous(limits = c(-0.45, 0.45)) +
      geom_point(size = 3)
    } else {
      plot_ordination(ordination = ord_result, 
                    physeq = phylist[[name]], # ADJUST TO PHYLOSEQ list name
                    color = "time", 
                    shape = "substrate",
                    title = paste(name, metric, sep = " - ")) +
      geom_point(size = 3)
    }
  })
  return(plots)
}

# Plot each ordination using the specified metric
metric_to_plot <- "Jac"  # Change this to "BC", "Jac", "UU", or "WU"
plots_Bray_Curtis <- plot_ordination_from_list(result_list, physeq_list_reps, metric = "bray")
plots <- plot_ordination_from_list(ordination_list, metric_to_plot)
gridExtra::grid.arrange(grobs = plots, ncol = 1)

# Display the plots
ggsave(plot = gridExtra::grid.arrange(grobs = plots, ncol = 1), filename =  glue("figures/{proj}_{metric_to_plot}_plots_all_in_one_ordination.pdf"), width = 7, height = 5)

# Plot each ordination using the specified metric
metric_to_plot <- "BC"  # Change this to "BC", "Jac", "UU", or "WU"
plots <- plot_ordination_from_list(ordination_list, metric_to_plot)
#gridExtra::grid.arrange(grobs = plots, ncol = 1)

# Display the plots
ggsave(plot = gridExtra::grid.arrange(grobs = plots, ncol = 1), filename =  glue("figures/{proj}_{metric_to_plot}_plots_all_in_one_ordination.pdf"), width = 7, height = 5)
```

#### Direct visualisation of beta diversity distances/dissimilarities between consecutive time points
```{r beta diversity distances within runs, eval=F, echo=F, fig.height=5, fig.width=7, message=FALSE, warning=FALSE, cache=FALSE, include=F}

# tax info
tax.df <- as.data.frame(tax_table(psdata_Genus_lowN))
tax.df$OTU <- rownames(tax.df)

# psmelt
psmelt_Genus_lowN =
psdata_Genus_lowN %>% 
  psmelt() %>% 
  select(Sample, OTU, everything())
  
# ASV table core community water
Genus_lowN_table = 
psmelt_Genus_lowN %>% 
  select(Sample, OTU, counts=Abundance) %>% 
  pivot_wider(names_from = "OTU", values_from = "counts", values_fill = 0) %>% 
  data.frame(row.names = 1) %>% as.matrix() %>% round()

# beta diversity

# avgdist Bray curtis dist matrix to use:
bray_Genus_lowN_table = vegdist(Genus_lowN_table, method = "bray")
jac_Genus_lowN_table = vegdist(Genus_lowN_table, method = "jaccard")
uu_Genus_lowN_table = UniFrac(psdata_Genus_lowN, weighted = F)
wu_Genus_lowN_table = UniFrac(psdata_Genus_lowN, weighted = T)

# reshape to long format
jac_Genus_lowN_long = 
jac_Genus_lowN_table %>% 
  as.matrix() %>% 
  as_tibble(rownames = "sample_a") %>% 
  pivot_longer(-sample_a, names_to = "sample_b", values_to = "distance") %>% 
  dplyr::mutate(metric = "jaccard")

bray_Genus_lowN_long = 
bray_Genus_lowN_table %>% 
  as.matrix() %>% 
  as_tibble(rownames = "sample_a") %>% 
  pivot_longer(-sample_a, names_to = "sample_b", values_to = "distance") %>% 
  dplyr::mutate(metric = "braycurtis")

uu_Genus_lowN_long = 
  uu_Genus_lowN_table %>% 
  as.matrix() %>% 
  as_tibble(rownames = "sample_a") %>% 
  pivot_longer(-sample_a, names_to = "sample_b", values_to = "distance") %>% 
  dplyr::mutate(metric = "unweighted unifrac")

wu_Genus_lowN_long = 
  wu_Genus_lowN_table %>% 
  as.matrix() %>% 
  as_tibble(rownames = "sample_a") %>% 
  pivot_longer(-sample_a, names_to = "sample_b", values_to = "distance") %>% 
  dplyr::mutate(metric = "weighted unifrac")
  
# define consecutive sampling moments
consecutive_timepoints = c("0h_to_24h", "24h_to_48h")

# reshape distance data and combine with metadata
all_dists_Genus_lowN =
  bind_rows(
    
bray_Genus_lowN_long %>% 
  filter(sample_a != sample_b) %>% # remove self-comparisons
  inner_join(., alpha2 %>% select(Sample,paper_id:time), by = c("sample_a" = "Sample")) %>% 
  rename_with(., .fn = ~paste0(.,"_a"), .cols = names(.)[-c(1:4)]) %>% # tag metadata to first sample
  inner_join(., alpha2 %>% select(Sample,paper_id:time), by = c("sample_b" = "Sample")) %>%
  rename_with(., .fn = ~paste0(.,"_b"), .cols = names(.)[c(13:ncol(.))]) %>% # tag metadata to second sample
  dplyr::mutate(consec_timepoints = paste0(time_a,"_to_", time_b)) %>%
  filter(consec_timepoints %in% consecutive_timepoints) %>% # select only distances between consecutive times 
  filter(sludge_source_a == sludge_source_b, # select only within location comparisons
         substrate_a == substrate_b, # select only within substrate feed comparisons
         replicate_a == replicate_b # select only within replicate comparisons
         ) %>% 
  dplyr::mutate(stability = 1-distance)
,
jac_Genus_lowN_long %>% 
filter(sample_a != sample_b) %>% # remove self-comparisons
  inner_join(., alpha2 %>% select(Sample,paper_id:time), by = c("sample_a" = "Sample")) %>% 
  rename_with(., .fn = ~paste0(.,"_a"), .cols = names(.)[-c(1:4)]) %>% # tag metadata to first sample
  inner_join(., alpha2 %>% select(Sample,paper_id:time), by = c("sample_b" = "Sample")) %>%
  rename_with(., .fn = ~paste0(.,"_b"), .cols = names(.)[c(13:ncol(.))]) %>% # tag metadata to second sample
  dplyr::mutate(consec_timepoints = paste0(time_a,"_to_", time_b)) %>%
  filter(consec_timepoints %in% consecutive_timepoints) %>% # select only distances between consecutive times 
  filter(sludge_source_a == sludge_source_b, # select only within location comparisons
         substrate_a == substrate_b, # select only within substrate feed comparisons
         replicate_a == replicate_b # select only within replicate comparisons
         ) %>% 
  dplyr::mutate(stability = 1-distance)
,
uu_Genus_lowN_long %>% 
  filter(sample_a != sample_b) %>% # remove self-comparisons
  inner_join(., alpha2 %>% select(Sample,paper_id:time), by = c("sample_a" = "Sample")) %>% 
  rename_with(., .fn = ~paste0(.,"_a"), .cols = names(.)[-c(1:4)]) %>% # tag metadata to first sample
  inner_join(., alpha2 %>% select(Sample,paper_id:time), by = c("sample_b" = "Sample")) %>%
  rename_with(., .fn = ~paste0(.,"_b"), .cols = names(.)[c(13:ncol(.))]) %>% # tag metadata to second sample
  dplyr::mutate(consec_timepoints = paste0(time_a,"_to_", time_b)) %>%
  filter(consec_timepoints %in% consecutive_timepoints) %>% # select only distances between consecutive times 
  filter(sludge_source_a == sludge_source_b, # select only within location comparisons
         substrate_a == substrate_b, # select only within substrate feed comparisons
         replicate_a == replicate_b # select only within replicate comparisons
         ) %>% 
  dplyr::mutate(stability = 1-distance)
,
wu_Genus_lowN_long %>% 
  filter(sample_a != sample_b) %>% # remove self-comparisons
  inner_join(., alpha2 %>% select(Sample,paper_id:time), by = c("sample_a" = "Sample")) %>% 
  rename_with(., .fn = ~paste0(.,"_a"), .cols = names(.)[-c(1:4)]) %>% # tag metadata to first sample
  inner_join(., alpha2 %>% select(Sample,paper_id:time), by = c("sample_b" = "Sample")) %>%
  rename_with(., .fn = ~paste0(.,"_b"), .cols = names(.)[c(13:ncol(.))]) %>% # tag metadata to second sample
  dplyr::mutate(consec_timepoints = paste0(time_a,"_to_", time_b)) %>%
  filter(consec_timepoints %in% consecutive_timepoints) %>% # select only distances between consecutive times 
  filter(sludge_source_a == sludge_source_b, # select only within location comparisons
         substrate_a == substrate_b, # select only within substrate feed comparisons
         replicate_a == replicate_b # select only within replicate comparisons
         ) %>% 
  dplyr::mutate(stability = 1-distance)
)


# plot distances
plot_distances_Genus_lowN = 
all_dists_Genus_lowN %>% 
  ggplot(aes(x = consec_timepoints, y = distance, fill = substrate_a, group = substrate_a)) +
    geom_col(aes(fill=substrate_a), position = position_dodge(1), show.legend = F) +
    #geom_point(size = 2, shape =21, color = "black", position = position_dodge(1), show.legend = T) + # position = position_jitterdodge(0.2)) +
    scale_color_manual(values = c(brewer.pal(6, "Dark2"))) +
    scale_fill_manual(values = c(brewer.pal(6, "Dark2"))) +
    scale_shape_manual(values = c(19,17, 2)) +
    scale_y_continuous(limits = c(0,1)) +
    #theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    theme(legend.text = element_markdown(),
          legend.key.size = unit(7, "pt"),
          axis.ticks.x = element_blank(), 
          #strip.background = element_rect(colour = "grey50"),
          strip.placement = "outside",
          legend.position = "right",
          panel.border = element_rect(colour = "grey50", fill = NA)) +
    facet_nested(rows = vars(metric), cols=vars(sludge_source_a, substrate_a, replicate_a), space = "free", scales = "free") +
    labs(y = "Distance \n(0 = identical composition)",
         x = NULL) 

ggsave(plot_distances_Genus_lowN, 
       filename = glue("figures/{proj}_community_distances_within_experiments.pdf"), width = 12, height = 9)

```