---
title: "YXIN_butyrate_feed_PHA"
author: "Pieter van Veelen"
date: "2023-10-25"
output: html_document
---

```{r , eval=TRUE, echo=T, include=T, message=FALSE, warning=FALSE}

# set optional parameters
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE, 
                      message = FALSE)
options(scipen = 999, digits = 3)

```

```{r library loading, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

## load required packages
library(Hmisc)
library(phyloseq)
library(qiime2R)
library(tidyverse)
library(magrittr)
library(devtools)
library(qiime2R)
library(here)
library(breakaway)
library(DivNet)
library(openxlsx)
library(ape)
library(vegan)
library(ggtext)
library(cowplot)
library(RColorBrewer)
library(microbiome)
library(lme4)
library(lmerTest)
library(decontam)
library(ampvis2)
library(glue)
library(lubridate)
library(DECIPHER)
library(ensembleTax)
library(mEQO)
library(ggh4x)

```

```{r project organization, message=F, echo=T, eval=T, warning=F, include=T, cache=T}

# project name
proj = "YXIN_proj1_Q16712"

# create directories
if(!dir.exists("figures")){dir.create("figures")}
if(!dir.exists("output_data")){dir.create("output_data")} 
if(!dir.exists("scripts")){dir.create("scripts")} 
if(!dir.exists("scripts/QIIME2")){dir.create("scripts/QIIME2")} 

### create color vector for later use
library(RColorBrewer)
colorset <- c("darkblue", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "lightskyblue", "darkgreen",  "khaki2", "firebrick", "brown1", "darkorange1", "cyan1", "royalblue4", "darksalmon", "darkblue", "royalblue4", "dodgerblue3", "steelblue1", "lightskyblue", "darkseagreen", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "brown1", "darkorange1", "cyan1", "darkgrey", "darkblue", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "lightskyblue", "darkgreen", "deeppink", "khaki2", "firebrick", "brown1", "darkorange1", "cyan1", "royalblue4", "darksalmon", "darkblue", "royalblue4", "dodgerblue3", "steelblue1", "lightskyblue", "darkseagreen", "darkgoldenrod1", "darkseagreen", "darkorchid", "darkolivegreen1", "brown1", "darkorange1", "cyan1", "darkgrey")


```


```{r import data, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

physeq = 
qza_to_phyloseq(
  features = "input_data/XYIN_16S_515F_926R_07092023_Q16712_XYIN_table.qza",
  tree = "input_data/XYIN_16S_515F_926R_07092023_Q16712_XYIN_rooted-tree.qza",
  taxonomy = "input_data/XYIN_16S_515F_926R_07092023_Q16712_XYIN_silva-138-99-nb-classifier.qza",
  metadata = "input_data/XYIN_16S_515F_926R_07092023_Q16712_XYIN@metadata_formatted.txt"
  )

# write raw phyloseq object as .rds object
write_rds(physeq, "output_data/physeq.rds")

# check sample metadata and column types
sample_data(physeq)

```

```{bash}

#cp -R /Users/pvee/Wetsus_Projects/BKIS/BKIS_proj1_Q16705/data_analysis/scripts ./scripts

```

```{r clean taxonomy, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

source("scripts/tax_clean_updated.R")

physeq_clean = tax_clean(physeq)

write_rds(physeq_clean, "output_data/physeq_clean_tax.rds")

```

```{r view non-bacterial sequences, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

psdata = physeq_clean
# clean non-informative and non-bacterial ASVs

# quick view on proportion non-bacterial sequences
psdata_nonBac =
psdata %>% 
  tax_glom(., "Phylum") %>% 
  transform_sample_counts(., function(x) x/sum(x)) %>% 
  psmelt %>% 
  as_tibble() %>% 
  select(Sample, Abundance, Kingdom, Phylum) %>% 
  filter(Kingdom != "Bacteria") %>% 
  group_by(Sample, Phylum) %>% 
  summarise(rel_abund = sum(Abundance)) 

psdata_nonBac %>% 
  ggplot(aes(x=Sample, y = rel_abund, fill = Phylum)) +
  geom_col() +
  scale_fill_manual(name=NULL, 
                    values = colorset) + # with named color vector ordered by median Genus abundance
  scale_y_continuous(expand = c(0,0)) +
  labs(x=NULL, y="Relative Abundance (%)") +
  theme_classic() +
  theme(axis.text.x = element_markdown(angle = 45, hjust = 1),
        legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        legend.position = "bottom")


# quick view on proportion Chloroplast and Mitochondrial sequences
psdata_mito_chlorop = 
psdata %>% 
  tax_glom(., "Family") %>% 
  transform_sample_counts(., function(x) x/sum(x)) %>% 
  psmelt %>% 
  as_tibble() %>% 
  filter(Kingdom == "Bacteria") %>% 
  select(Sample, Abundance, Kingdom, Phylum, Family) %>% 
  filter(grepl("Mitochondria|Chloroplast", Family)) %>% 
  group_by(Sample, Family) %>% 
  summarise(rel_abund = sum(Abundance))

psdata_mito_chlorop %>% 
  ggplot(aes(x=Sample, y = rel_abund, fill = Family)) +
  geom_col() +
  scale_fill_manual(name=NULL, 
                    values = colorset) + # with named color vector ordered by median Genus abundance
  scale_y_continuous(expand = c(0,0)) +
  labs(x=NULL, y="Relative Abundance (%)") +
  theme_classic() +
  theme(axis.text.x = element_markdown(angle = 45, hjust = 1),
        legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        legend.position = "bottom")

```

```{r clean on taxonomy, message=F, echo=T, eval=T, warning=F, include=T, cache=F}

# filter unwanted taxonomic groups out
psdata_tax_filtered =
psdata %>% # 5531
  subset_taxa(.,grepl("Bacteria", Kingdom)) %>% # 21 ASVs lost because of other Kingdom than Bacteria
  subset_taxa(., Kingdom != "Unassigned") %>% # 0 ASVs lost
  subset_taxa(., !is.na(Phylum)) %>% # 0 ASVs lost 
  subset_taxa(., Phylum != "Phylum of Bacteria") %>% # 118 ASVs lost
  subset_taxa(., Family != "Mitochondria") %>% # 65 ASVs lost
  subset_taxa(., Class != "Chloroplast") %>% # 0 ASVs lost
  subset_taxa(., Order != "Chloroplast") # 12 ASVs lost


# remove non-informative taxa
old = ntaxa(psdata)
new = ntaxa(psdata_tax_filtered)

# number of ASVs removed:
old-new # 216 ASVs are removed

# check removed Mitochondrial ASV counts per sample
psdata %>% 
  subset_taxa(., Family == "Mitochondria") %>% 
  psmelt %>% 
  select(OTU, Sample, Family,Abundance) %>% 
  mutate_if(is.character, as.factor) %>% 
  group_by(Sample) %>%
summarise(Abundance=sum(Abundance)) %>% 
arrange(desc(Abundance), .by_group = T)

# save unfiltered psdata
psdata_unfiltered = psdata 
saveRDS(object = psdata_unfiltered, file = "output_data/psdata_unfiltered.rds") # save unfiltered data
psdata = psdata_tax_filtered # continue with filtered data as psdata

```

```{r clean phylogeny, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

### resolve phylogenetic tree ###
physeq = physeq_clean

# evaluate tree topology
is.binary(phy_tree(psdata_tax_filtered)) # if FALSE --> polychotomy present (node has more than 2 tips)
#TRUE

# # if FALSE:
# # resolve polychotomous nodes
phy_tree_resolved <- multi2di(phy_tree(physeq))
is.binary(phy_tree_resolved)
# create new phy_tree
tree2 <- phy_tree_resolved

# subset_taxa(phy_tree_resolved, Kingdom ==  "Bacteria")

# merge new phy_tree object with sample_data and otu_table into new phyloseq object
physeq2 <- merge_phyloseq(otu_table(physeq), sample_data(physeq), tax_table(physeq), tree2)

# now filter only the taxa that are clean_filtered by taxonomic filtering
psdata = subset_taxa(physeq2, taxa_names(physeq2) %in% taxa_names(psdata_tax_filtered))

# save filted phyloseq object with resolved tree
saveRDS(object = psdata, file = "output_data/psdata_filtered_tree-resolved.rds")

```

```{r rarefaction curves, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

# alpha rarefaction curve
source("scripts/ampvis2_internals.r")
source("scripts/amp_rankabundance.r")
source("scripts/amp_rarecurve.r")

# show plot
amp_rarecurve(psdata, color = "exp_id", legend.position = "bottomright", xlim = c(0, 120000))

# save plot
pdf(glue("figures/{proj}_rarefaction_curves.pdf"), useDingbats = F, width = 7, height = 5)
amp_rarecurve(psdata, color = "exp_id", xlim = c(0, 120000), legend.position = "bottomright")
dev.off()


```

```{r view mock and blank, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

# extract DNA extraction blank # 216 ASVs
psdata_blank = 
  prune_taxa(taxa_sums(
    subset_samples(psdata, exp_id  == "blank_nc")) > 0,
    subset_samples(psdata, exp_id  == "blank_nc")
    )

# extract mock community # 287 ASVs
psdata_mock = 
  prune_taxa(taxa_sums(
    subset_samples(psdata, exp_id  == "mock")) > 0,
    subset_samples(psdata, exp_id  == "mock")
    )

saveRDS(object = psdata_mock, file = "~/Wetsus_Projects/PVEE/Data_analyses/mock_communities_wetsus/mock_communities_wetsus/input_data/YXIN_proj1_Q16712_PHA_butyrate_feed_mock.rds")

```

```{r remove_mock_and_nc}

# keep samples only # 5246 ASVs
psdata_samples = 
  prune_taxa(taxa_sums(
    subset_samples(psdata, !grepl("mock|blank_nc", exp_id))) >0,
    subset_samples(psdata, !grepl("mock|blank_nc", exp_id))
    )

# recreate psdata with samples only
psdata = psdata_samples

```

```{r filter on abundance, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

# relative abundance data
psdata_rel <- transform_sample_counts(psdata, fun = function(x) x/sum(x)) # psdata_rel taxa
ntaxa(psdata_rel)

# total sequence sum
total_sum = sum(sample_sums(psdata)) # 1859672 reads

# continue downstream analysis with abundance filter that retains ASVs with at least 0.1% of total read abundance -> keeps above 99% abundance when removing rare data
psdata_0.01pct <- prune_taxa(taxa_sums(psdata_rel) > 0.0001, psdata) # 3601 taxa (99.7% abundance)
psdata_0.05pct <- prune_taxa(taxa_sums(psdata_rel) > 0.0005, psdata) # 2107 taxa (98.6% abundance)
psdata_0.1pct <- prune_taxa(taxa_sums(psdata_rel) > 0.001, psdata) # 1607 taxa (97.4% abundance)
psdata_1pct <- prune_taxa(taxa_sums(psdata_rel) > 0.01, psdata) # 492 taxa (84.9% abundance)

sum(sample_sums(psdata_0.01pct))/total_sum*100
sum(sample_sums(psdata_0.05pct))/total_sum*100
sum(sample_sums(psdata_0.1pct))/total_sum*100
sum(sample_sums(psdata_1pct))/total_sum*100

psdata_NoAbundanceFilter = psdata
psdata <- psdata_0.05pct # keeping 98.6% of reads in 2107 ASVs

# difference in sample coverage (1.96)
max(sample_sums(psdata_0.05pct)/min(sample_sums(psdata_0.05pct)))

```

```{r prepare data for picrust2, eval=F}

# input data = psdata_0.05pct
biom_test =
psdata_0.05pct %>% 
  otu_table() %>% 
  data.frame(rownames = 1) %>% 
  as_tibble()

biomformat::write_biom(biom_test, glue("output_data/{proj}_otu_table_psdata_0.05pct.biom"))

# try running PICRUST2 as QIIME2 plugin
# DONE

```

```{r agglomerate to genus level}

# psdata Genus 
psdata_Genus = tax_glom(
  transform_sample_counts(
    psdata, 
    function(x) x/sum(x)), "Genus")

# psdata Genus Paper 1: Low N conditions
psdata_Genus_lowN = 
  tax_glom(
    prune_taxa(
      taxa_sums(
        subset_samples(psdata, paper_id == "first"))>0, 
        subset_samples(psdata, paper_id == "first")), "Genus")

# psdata Genus Paper 2: High N conditions
psdata_Genus_highN = 
  tax_glom(
    prune_taxa(
      taxa_sums(
        subset_samples(psdata, paper_id == "second"))>0, 
        subset_samples(psdata, paper_id == "second")), "Genus")

# create relative abundance phyloseq objects
psdata_Genus_lowN_rel = transform_sample_counts(psdata_Genus_lowN, function(x) x/sum(x))
psdata_Genus_highN_rel = transform_sample_counts(psdata_Genus_highN, function(x) x/sum(x))


# psmelts Genus
ps_Genus = psmelt(psdata_Genus) # n=426
ps_Genus_lowN = psmelt(psdata_Genus_lowN) # n= 426
ps_Genus_highN = psmelt(psdata_Genus_highN) # n= 357

```

```{r create PDT input Genus}

# create absolute genus physeq
# psdata Genus 
psdata_Genus_abs = tax_glom(psdata, "Genus")

long = 
psdata_Genus_abs %>% 
  psmelt() %>% 
  as_tibble()

# calculate genus-level taxa_sums across samples and calculate taxa prevalence %
wide_taxsums_prev = long %>% 
  select(Sample, Kingdom:Genus, Abundance) %>% 
  pivot_wider(names_from = "Sample", values_from = "Abundance") %>% 
  rowwise() %>% 
  mutate(taxa_sums = sum(c_across(any_of(starts_with("YXIN")))),
         prevalence = mean(c_across(any_of(starts_with("YXIN"))) != 0, na.rm = TRUE) * 100) %>% 
  select(Kingdom:Genus, prevalence, taxa_sums, everything()) 

# make data long again
long_taxsums_prev = 
wide_taxsums_prev %>%
  pivot_longer(cols = any_of(starts_with("YXIN")), names_to = "Sample", values_to = "Abundance")

# calculate sample_sums for sample set
long_sampsums = 
wide_taxsums_prev %>% 
  pivot_longer(cols = any_of(starts_with("YXIN")), names_to = "Sample", values_to = "Abundance") %>% 
  select(Genus, Sample, Abundance) %>% 
  pivot_wider(names_from = "Genus", values_from = "Abundance") %>% 
  rowwise() %>% 
  mutate(sample_sums = sum(c_across(2:ncol(.)))) %>% 
  select(Sample, sample_sums, everything()) %>% 
  pivot_longer(cols = c(3:ncol(.)), names_to = "Genus", values_to = "Abundance")

## PDT data formatted

# which prevalence and abundance thresholds in (%)
# vectors should be of same lengths corresponding to the combinations to make
prevalences = c(80, 90)
abundances = c(0.5, 0.1)

# loop over the data applying combinations of prevalences
for(prev in prevalences) {
  for(abun in abundances) {
    
inner_join(long_sampsums, long_taxsums_prev, by = c("Sample", "Genus")) %>% 
  select(-Abundance.y) %>% 
  rename(Abundance.x = "Abundance") %>% 
  mutate(perc_abund = Abundance/sample_sums*100) %>% 
  select(Sample, Genus, perc_abund, prevalence, "genus_sums" = taxa_sums, Family) %>% 
  group_by(Family, Genus, prevalence, genus_sums) %>% 
  summarise(mean_perc_abund = mean(perc_abund), max_perc_abund = max(perc_abund), min_perc_abund = min(perc_abund), .groups = "drop") %>% 
  filter(prevalence >= prev, mean_perc_abund >= abun) %>% 
  arrange(desc(max_perc_abund)) %>% 
  mutate(`0_Index` = row_number(),
         `1_Name` = if_else(grepl("Genus of ", Genus), Family, Genus),
         `2_Tax` = if_else(grepl("Genus of ", Genus), "family", "genus"),
         `3_mean_genus_perc` = mean_perc_abund,
         `4_perc_samples` = prevalence) %>% 
  select(`0_Index`, `1_Name`, `2_Tax`, `3_mean_genus_perc`, `4_perc_samples`, Family) %>% 
  arrange(desc(`2_Tax`)) %>% 
  write_csv(glue::glue("output_data/{proj}_PDT_input_Genus_prev{prev}_abund{abun}pct.csv"), col_names = T)
  }
}


```


```{r alpha diversity rarefy}

source("scripts/avgrarefy.r")

# create subset frequency matrix
YXIN_matrix_lowN <- as(t(otu_table(psdata_Genus_lowN)), "matrix")
YXIN_matrix_highN <- as(t(otu_table(psdata_Genus_highN)), "matrix")

# determine minimal sampling depth
min_sample_lowN <- min(sample_sums(psdata_Genus_lowN)) # 42061
min_sample_highN <- min(sample_sums(psdata_Genus_highN)) # 50627


# rarefaction taking mean of 100 iterations
set.seed(711)
YXIN_matrix_lowN_rare42061_table = avgrarefy(x=YXIN_matrix_lowN, sample = min_sample_lowN, iterations = 100, seed = 711)
YXIN_matrix_highN_rare50627_table = avgrarefy(x=YXIN_matrix_highN, sample = min_sample_highN, iterations = 100, seed = 711)


# create phyloseq object with rarefied data
# low N
psdata_YXIN_lowN_rare <- psdata_Genus_lowN
otu_rare_lowN = otu_table(data.frame(t(YXIN_matrix_lowN_rare42061_table)), taxa_are_rows = TRUE)
otu_table(psdata_YXIN_lowN_rare) <- otu_rare_lowN

# high N
psdata_YXIN_highN_rare <- psdata_Genus_highN
otu_rare_highN = otu_table(data.frame(t(YXIN_matrix_highN_rare50627_table)), taxa_are_rows = TRUE)
otu_table(psdata_YXIN_highN_rare) <- otu_rare_highN

```

### All data are ready for analysis of both papers
Below for now only microbial diversity analysis of Paper 1: low N samples

#### Alpha Diversity

```{r plot alpha diversity rare, message=F, echo=T, eval=T, warning=T, include=T, cache=F}

# calculate alpha diversity: richness & Shannon on Genus level
# input = 
psdata_YXIN_lowN_rare 

# rarefied
alpha <- estimate_richness(psdata_YXIN_lowN_rare, measures = c("Observed", "Chao1", "Shannon", "Simpson"))
alpha$Sample <- row.names(alpha)

metadata = 
  sample_data(psdata_YXIN_lowN_rare) %>% 
  data.frame() %>% 
  as_tibble() %>% 
  mutate(Sample = sample_names(psdata_YXIN_lowN_rare)) %>% 
  select(Sample, everything()) 

# join alpha diversity with metadata
alpha2 <- inner_join(metadata, alpha, by = "Sample") #%>% 
  #mutate(biofilm = factor(biofilm, levels = c("Broth", "Biofilm")))

# plot alpha diversity
Chao1 <- 
alpha2 %>% 
ggplot(aes(x=time, y=Chao1, color = substrate, fill = substrate)) +
  #stat_summary(fun = median, geom = "bar", alpha = 0.3) +
  geom_point(aes(#shape = sampling_day, 
                 fill = substrate,
                 color = substrate),
                 #position = position_jitter(width = 0.3)
                 , size = 2, show.legend = T) + 
    geom_line(aes(group = substrate, 
                 color = substrate),
                 , linewidth = 1,alpha=0.5, show.legend = F) + 
  scale_color_manual(values = brewer.pal(6, "Paired")) +
  scale_fill_manual(values = brewer.pal(6, "Paired")) +
  scale_y_continuous(limits = c(0,NA)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(#legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        #axis.ticks.x = element_blank(), 
        strip.background = element_rect(fill="white", colour = "white"),
        axis.title.y = element_text(hjust = 0.5),
        strip.placement = "outside",
        legend.position = "bottom",
        panel.border = element_rect(colour = "black", fill = NA)) +
  labs(y = expression("Genus Chao1"),
       x = NULL) +
  theme(axis.text.x = element_text(hjust = 0.5)) +
  facet_nested(~ sludge_source + replicate, scales = "free", space = "free_x")


# print Chao1 Genus
print(Chao1)



# plot alpha diversity
Shannon <- 
alpha2 %>% 
ggplot(aes(x=time, y=Shannon, color = substrate, fill = substrate)) +
  #stat_summary(fun = median, geom = "bar", alpha = 0.3) +
  geom_point(aes(#shape = sampling_day, 
                 fill = substrate,
                 color = substrate),
                 #position = position_jitter(width = 0.3)
                 , size = 2, show.legend = T) + 
    geom_line(aes(group = substrate, 
                 color = substrate),
                 , linewidth = 1,alpha=0.5, show.legend = F) + 
  scale_color_manual(values = brewer.pal(6, "Paired")) +
  scale_fill_manual(values = brewer.pal(6, "Paired")) +
  scale_y_continuous(limits = c(0,NA)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(#legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        #axis.ticks.x = element_blank(), 
        strip.background = element_rect(fill="white", colour = "white"),
        axis.title.y = element_text(hjust = 0.5),
        strip.placement = "outside",
        legend.position = "bottom",
        panel.border = element_rect(colour = "black", fill = NA)) +
  labs(y = expression("Genus Shannon"),
       x = NULL) +
  theme(axis.text.x = element_text(hjust = 0.5)) +
  facet_nested(~ sludge_source + replicate, scales = "free", space = "free_x")

print(Shannon)

# create plot
prow = 
  plot_grid(
  plot_grid(Chao1 + theme(legend.position = "none"), 
            Shannon + theme(legend.position = "none"),
                   align = "hv",
                   labels = c("A", "B"),
                   hjust = -1, 
                   nrow = 2),
  get_legend(Chao1), 
  nrow = 2 , 
  rel_heights = c(5, 1))


# show plot
prow

# save plot
ggsave(prow, filename = glue("figures/{proj}_plot_alpha_div.pdf"), width = 7, height = 5)
 

```

```{r write genus tables}

# paper 1 only
### write rel abund data plus other tax info
ps_Genus_lowN %>% 
  # filter(!grepl("mock", combined), # remove mock communities
  #        Sample_mock != "blank") %>% # remove NC blanks
  as_tibble() %>% 
  mutate(across(everything(),  ~str_replace_all(., ",", "."))) %>% 
  mutate(across(where(~ all(grepl("^-?\\d*\\.?\\d*$", .))), as.numeric)) %>% 
  select(Sample, OTU, Abundance, Kingdom:Genus, everything()) %>% 
  group_by_at(vars(Sample, Kingdom:Genus)) %>% 
  summarise(rel_abund = sum(Abundance), .groups = "drop") %>% 
  group_by_at(vars(-rel_abund)) %>% 
  summarise(mean_rel_abund = 100* mean(rel_abund), .groups = "drop") %>% 
  select(Kingdom:Genus, Sample, mean_rel_abund) %>% 
  pivot_wider(names_from = Sample, values_from = mean_rel_abund, values_fill = 0) %>% 
  group_by_at(vars(Kingdom:Genus)) %>% 
  rowwise() %>% 
  mutate(mean = mean(c_across(where(is.numeric)))) %>% 
  arrange(desc(mean)) %>% 
  select(-mean) %>% 
  ungroup() %>% 
  mutate(across(where(is.numeric), ~round(.x, digits = 2))) %>% 
  select(Kingdom:Genus, everything()) %>% 
  write_csv(glue::glue("output_data/{proj}_relative_Genus_abundances_sorted_by_overall_mean.csv", col_names = T, append = F))

```

```{r plot_bar all genera, fig.width=8, fig.height=8}

# paper 1 only
#ps_Genus_lowN

# count n genera n = 426
ps_Genus_lowN %>% 
  as_tibble() %>% 
  mutate(Genus = factor(Genus)) %>% 
  pull(Genus) %>% 
  levels() %>% 
  length()

# Define a function to find the first non-matching column

# Use rowwise and mutate to apply the function
ps_Genus_taxlabel = ps_Genus_lowN %>% 
  as_tibble() %>% 
  mutate(across(everything(),  ~str_replace_all(., ",", "."))) %>% 
  mutate(across(where(~ all(grepl("^-?\\d*\\.?\\d*$", .))), as.numeric)) %>% 
  select(Sample, Abundance, paper_id:Genus) %>%
  rowwise() %>%
  mutate(tax_label = case_when(
    grepl("\\d", Genus) ~ {
      first_non_numeric <- case_when(
        !grepl("\\d", Family) ~ paste(Family, Genus, sep = " "),
        !grepl("\\d", Order) ~ paste(Order, Genus, sep = " "),
        !grepl("\\d", Class) ~ paste(Class, Genus, sep = " "),
        !grepl("\\d", Phylum) ~ paste(Phylum, Genus, sep = " "),
        !grepl("\\d", Kingdom) ~ paste(Kingdom, Genus, sep = " "),
        TRUE ~ NA_character_
      )
      first_non_numeric
    },
    TRUE ~ Genus
  ))

Genus_abundances <- ps_Genus_taxlabel %>% 
  mutate(Sample_nr = str_sub(Sample, start = 1, end = 2)) %>% 
  group_by_at(vars(Sample, Sample_nr, sludge_source, substrate, replicate,exp_id, time, tax_label)) %>% 
  summarise(rel_abund = sum(Abundance), .groups = "drop") %>%
  group_by(Sample) %>% 
  mutate(mean_rel_abund = rel_abund/sum(rel_abund)*100) %>% 
  ungroup() %>% 
  group_by(Sample, tax_label) %>% 
  mutate(tax_label = str_replace(tax_label, "(.*)_unclassified", "Unclassified *\\1*"),
         tax_label = str_replace(tax_label, "^(\\S*)$", "*\\1*"))

# Reduce genera to lower number (at least 3%, or else called other)
Genus_pool <- Genus_abundances %>% 
  group_by(tax_label) %>% 
  summarise(pool = max(mean_rel_abund) < 3, 
            mean = mean(mean_rel_abund), 
            .groups = "drop")

# define a color vector using brewer.pal and colorset
colvec = c(brewer.pal(8,"Dark2"),brewer.pal(8,"Set2"),brewer.pal(8,"Accent"),brewer.pal(8,"Paired"),brewer.pal(12,"Set3"),brewer.pal(9,"Set1"), brewer.pal(9,"YlGnBu"), colorset)

# name the taxon colors
inner_join(Genus_abundances, Genus_pool, by="tax_label") %>% 
  mutate(tax_label = if_else(pool, "Other max.<3%", tax_label)) %>% 
  group_by(tax_label) %>%
  summarise(mean_rel_abund = sum(mean_rel_abund), 
            median = median(mean),
            .groups = "drop") %>% 
  mutate(tax_label = factor(tax_label), 
         tax_label = fct_reorder(tax_label, median, .desc = T)) %>% 
  arrange(desc(median)) %>% pull(tax_label) %>% unique() -> names(colvec)

colvec["Other max.<3%"] <- "#D3D3D3"

# reshape plot data
plot_data <-
inner_join(Genus_abundances, Genus_pool, by="tax_label") %>% 
  mutate(tax_label = if_else(pool, "Other max.<3%", tax_label)) %>% 
  # mutate(Sample = if_else(grepl(".*[^0-9][0-9]{2}$", Sample),
  #                          Sample,
  #                          str_c(str_sub(Sample, end = -2),"0", str_sub(Sample, -1)))) %>% 
  group_by(Sample, Sample_nr, sludge_source, replicate, substrate, time, tax_label) %>% 
  summarise(mean_rel_abund = sum(mean_rel_abund), 
            mean = min(mean),
            .groups = "drop") %>% 
  mutate(tax_label = factor(tax_label), 
         tax_label = fct_reorder(tax_label, mean, .desc = T)) 



# plot the data per sample
plot_Genus = 
  plot_data %>% 
ggplot(aes(x=time, y = mean_rel_abund, fill = tax_label)) +
  geom_col() +
  scale_fill_manual(name=NULL, 
                    values = colvec) + # with named color vector ordered by median Genus abundance
  scale_y_continuous(expand = c(0,0)) +
  labs(x=NULL, y="Relative Abundance (%)") +
  theme_classic() +
  theme(axis.text.x = element_markdown(angle = 0, hjust = 0.5),
        legend.text = element_markdown(),
        legend.key.size = unit(7, "pt"),
        legend.position = "bottom",
        strip.background = element_rect(colour = "white"),
        strip.text = element_text(face = "bold")) +
  guides(fill = guide_legend(nrow=12))
  

# print plot  
gg_Genus = plot_Genus + facet_nested_wrap(~sludge_source + replicate + substrate, ncol=4,  scales = "free")
gg_Genus

# save plots
ggsave(plot = gg_Genus, filename = glue("figures/{proj}_barplot_all_Genus_reactorsOnly.pdf"), width = 8, height = 6)

```

```{r geom_tile}

# remarks order for x axis
#order = c(1,19,16,6,5,9,17,7,8,4,3,18,2,10,11,12,13,14,15)

taxon_pool <- Genus_abundances %>% 
  group_by(tax_label) %>% 
  summarise(pool = max(mean_rel_abund) < 3, 
            mean = mean(mean_rel_abund), 
            .groups = "drop")

heatmap_data_long = 
Genus_abundances %>% 
  group_by(Sample,sludge_source, replicate, substrate,time, tax_label) %>% 
  summarise(mean_rel_abund = mean(mean_rel_abund), .groups = "drop") %>% 
inner_join(., Genus_pool, by="tax_label") %>% 
  group_by(Sample,sludge_source, replicate, substrate,time, tax_label) %>% 
  mutate(tax_label = if_else(pool, "Other max.<3%", tax_label)) %>% 
  select(Sample, sludge_source, replicate, substrate, time,tax_label, mean_rel_abund, mean, pool) %>% 
  group_by(Sample, sludge_source, replicate, substrate, time,tax_label) %>% 
  summarise(sum_rel_abund = sum(mean_rel_abund)) %>% 
  mutate(tax_label = factor(tax_label), 
         tax_label = fct_reorder(tax_label, sum_rel_abund, .desc = F)) %>%
  ungroup()

# determine order of taxa on y axis from mean high to low
tax_order = 
heatmap_data_long %>% 
  mutate(tax_label = as.character(tax_label)) %>% 
  group_by(tax_label) %>% 
  summarise(mean = mean(sum_rel_abund)) %>% 
  arrange(desc(mean)) %>% pull(tax_label)

# create wide format data
heatmap_data = 
heatmap_data_long %>% 
select(Sample, sludge_source, replicate, substrate,time, tax_label,time, sum_rel_abund) %>% 
  pivot_wider(names_from = tax_label, 
              values_from = sum_rel_abund) 

# make heatmap
m <- as.matrix(heatmap_data[, 6:ncol(heatmap_data)])
clust <- hclust(dist(t(m)))

# plot heatmap
heatmap_genus_3pct = 
heatmap_data_long %>% 
  mutate(tax_label = fct_relevel(tax_label, rev(tax_order))) %>% 
ggplot(aes(time, tax_label) ) +
  geom_tile(aes(fill = sum_rel_abund), color = NA) +
  labs(x = NULL, y=NULL) +
  #scale_y_discrete(limits = colnames(m)[rev(clust$order)]) +
  scale_fill_gradient(low = "white", high = "darkred",
                      name = "Relative\nAbundance (%)\n(\u00B7 <1%)") +
  theme(axis.text.x = element_text(size = 8, angle = 90, hjust = 1, vjust = 0.5),
        strip.text = element_text(angle = 0),
        axis.ticks = element_blank(),
        axis.text.y = element_markdown(size = 8),
        plot.background = element_blank(),
        panel.border = element_rect(colour = "white", fill = NA)) +
  geom_text(
  aes(label = ifelse(tax_label != "Other max.<3%" & sum_rel_abund > 3, round(sum_rel_abund, digits = 0), "")),
  color = "grey50", size = unit(3, "pt")) +
  geom_text(
  aes(label = ifelse(tax_label != "Other max.<3%" & sum_rel_abund < 1, ".", "")),
  color = "grey80", size = unit(5, "pt"), position = position_nudge(y=0.5)) +
  facet_nested(~sludge_source  + replicate + substrate, space = "free", scales = "free")

print(heatmap_genus_3pct)

ggsave2(plot = heatmap_genus_3pct + 
                ggtitle(label = "Genus heatmap -\nYXIN: PHA - acetate vs. butyrate under low N", 
                        subtitle = "Genera with abundance >3% in at least one sample\nPercent abundance shown in cells when >3%"), 
        filename = glue("figures/{proj}_heatmap_genus_3pct.pdf"), width = 10, height = 8)

# save plot data with max.<3% others
heatmap_data_long %>% 
  mutate(tax_label = fct_relevel(tax_label, rev(tax_order))) %>% 
  select(Sample, tax_label, sum_rel_abund) %>%
  pivot_wider(names_from = "Sample", values_from = "sum_rel_abund", values_fill = 0) %>%
write.xlsx(., file = glue("output_data/{proj}_heatmap_data_genus_table_3pct.xlsx"), col_names = T)

# save plot data without other category
ps_Genus %>% 
  as_tibble() %>% 
  select(sludge_source, replicate, substrate, Sample,time, Abundance, Kingdom:Genus) %>% 
  group_by_at(vars(sludge_source, replicate, substrate, Sample,time, Kingdom:Genus)) %>% 
  summarize(rel_abund = sum(Abundance), .groups = "drop") %>% 
  group_by_at(vars(Sample, Kingdom:Genus)) %>% 
  summarize(mean_rel_abund = 100* mean(rel_abund), .groups = "drop") %>% 
  mutate(Genus = str_replace(Genus, "(.*)_unclassified", "Unclassified *\\1*"),
         Genus = str_replace(Genus, "^(\\S*)$", "*\\1*")) %>% 
  pivot_wider(names_from = "Sample", values_from = "mean_rel_abund", values_fill = 0) %>%
  write.xlsx(., file = glue("output_data/{proj}_heatmap_data_genus_table_all.xlsx"), col_names = T) 
  
```

### Beta diversity

```{r total beta diversity rel_abund, message=F, echo=T, eval=F, warning=F, include=T, cache=F}

# recode days as categorical
colnames(sample_data(psdata_Genus_lowN))
sample_data(psdata)$sampling_day = as.factor(sample_data(psdata)$sampling_day)

### Beta diversity analysis first
# input data  = psdata
psdata_Genus_lowN_rel = transform_sample_counts(psdata, function(x) x/sum(x))
psdata_Genus_lowN_clr = microbiome::transform(psdata_Genus_lowN, transform = "clr")

sample_data(psdata_Genus_lowN_rel)$replicate = factor(sample_data(psdata_Genus_lowN_rel)$replicate)

# ordination
PCoA_BC <- ordinate(psdata_Genus_lowN_rel, method = "PCoA", distance = "bray")
PCoA_Jac <- ordinate(psdata_Genus_lowN_rel, method = "PCoA", distance = "jaccard")
PCoA_uu <- ordinate(psdata_Genus_lowN_rel, method = "PCoA", distance = "uunifrac")
PCoA_wu <- ordinate(psdata_Genus_lowN_rel, method = "PCoA", distance = "wunifrac")

nmds_BC <- ordinate(psdata_Genus_lowN_rel, method = "NMDS", distance = "bray")
nmds_Jac <- ordinate(psdata_Genus_lowN_rel, method = "NMDS", distance = "jaccard")
nmds_uu <- ordinate(psdata_Genus_lowN_rel, method = "NMDS", distance = "uunifrac")
nmds_wu <- ordinate(psdata_Genus_lowN_rel, method = "NMDS", distance = "wunifrac")

plot_PCoA_BC <- 
  plot_ordination(physeq = psdata_Genus_lowN_rel, 
                  ordination = nmds_BC, 
                  type = "samples", 
                  axes = c(1,2), 
                  color = "time", shape = "replicate") + 
  ggrepel::geom_text_repel(aes(label = Description), size = 2) +
  scale_color_manual(values = c(brewer.pal(6, "Dark2"))) + 
  geom_point(size=2) +
  ggtitle("Bray Curtis\n(presence and abundance)") + 
  facet_nested(~sludge_source+substrate)
  

plot_PCoA_Jac <- 
  plot_ordination(physeq = psdata_Genus_lowN_rel, 
                  ordination = PCoA_Jac, 
                   type = "samples", 
                  axes = c(1,2), 
                  color = "time", shape = "replicate") + 
  ggrepel::geom_text_repel(aes(label = Description), size = 2) +
  scale_color_manual(values = c(brewer.pal(6, "Dark2"))) + 
  geom_point(size=2) +
  ggtitle("Jaccard\n(binary presence only)") +
  facet_nested(~sludge_source + substrate)

plot_PCoA_uu <- 
  plot_ordination(physeq = psdata_Genus_lowN_rel, 
                  ordination = nmds_uu, 
                   type = "samples", 
                  axes = c(1,2), 
                  color = "time", shape = "replicate") + 
  ggrepel::geom_text_repel(aes(label = Description), size = 2) +
  scale_color_manual(values = c(brewer.pal(6, "Dark2"))) + 
  geom_point(size=2) +
  ggtitle("unweighted UniFrac \n(binary lineage presence only)") +
  facet_nested(~sludge_source + substrate)


# Combine the plots with shared legend
combined_plots <- patchwork::wrap_plots(#plot_PCoA_Jac, 
                                        plot_PCoA_BC, plot_PCoA_uu) +
  patchwork::plot_layout(guides = "collect", ncol = 1) 

# Display the combined plot
print(combined_plots)

ggsave(plot = combined_plots, filename = glue("figures/{proj}_plot_beta_diversity_all_clr.pdf"), width = 9, height =6)


```